{
 "cells": [
  {
   "attachments": {
    "agama.jpg": {
     "image/jpeg": "/9j/4AAQSkZJRgABAQEASABIAAD/2wBDAAUDBAQEAwUEBAQFBQUGBwwIBwcHBw8LCwkMEQ8SEhEPERETFhwXExQaFRERGCEYGh0dHx8fExciJCIeJBweHx7/2wBDAQUFBQcGBw4ICA4eFBEUHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh4eHh7/wAARCADIAtADASEAAhEBAxEB/8QAHAABAAIDAQEBAAAAAAAAAAAAAAcIBAUGAwIB/8QARxAAAQMDAgMFBgMECQIGAgMAAQACAwQFEQYSByExCBNBUWEUIjJxgZEVobEjQlLBFiQzYnKCktHwF+ElQ1OisvE3wmNk4v/EABsBAQADAQEBAQAAAAAAAAAAAAABAgMFBAYH/8QANxEBAAICAQQBAgMHAgQHAAAAAAECAxEEBRIhMUETUSIyYQYUQnGBobEVIySR0fA0NVJyweHx/9oADAMBAAIRAxEAPwC5aICICICINDqHVlpsdyprdWOldU1Dd4ZE0Ha3ONzskYHX7Fb5TMTEbBFAIgIgIgIgIgL5keyNhfI9rGjqXHACDR3DWWlrfWzUVZfaKKpgbuliMmXNGM9B6LWf9T9DNEQl1DTMkkaDs2vOMjP8KtFLSNxZtWaYvLmstWoLZWSO+GOKpY55/wAuc/ktyomJj2CKARARARARARAXyySN7nNZIxzmHDgDktPqg+kQEQEQEQEQEQEQEQEQEQEQEQEQEQEQEQEQEQEQEQEQEQEQEQEQEQEQF419VDQ0NRW1L9kFPE6WR3k1oyT9ggqvrbW9VdNSVF4MQikmh2wBjnENjHweHj1+6sJwqvsV90Vb5xOZZ44WMmJHPOOR+oW2Sv4UQ6pFikRARARARAXxNLFCzfLIyNv8TnABBXjirx/qYbhW2PRMbRLT+6a9zGyHd/cY7l1yPe8lD9/1Vr/U++kr73dKjfhz4y8RRudjHNjMDpy9VvFIiNqzLZaZ4c6yrIqqKno5+9ih71/egxggvAdh/L5/RfWpeHmvaAQsbY6+qo2sJ76KMvAHTGfLy81f8O/adRpzFHQ3yy2P8R76anhFS+ISMeQ+KSPHUDBGN45+pCnXh3xqrKKqfbtU1kNc1r2x+0t6uwdriPD154VbVi0I3pYC11tPcrbTXGkf3lPUxNmid5tcAR+RWQvOsIgIgIgIgIg+J5Y4IJJ5XBkcbS97j4ADJKr3pDi7bW8TaoPhqXUVaXQxua3cWMGXtkc1v1HyatMdZtuIRKw4IIBByCizSIgIgIgIg0+trnNZtK190pwHS00e9o8+YXL8KOJNLrQT0z4O4q4T0HwvHP8A2VteNo+UgIqpEQEQEQFpdZ6ltmlbJNc7nUCJrWnu27S50jgMgADmVMRsZGl71R6hsNJeaASCnqmlzBI3a4YJBBHzBWyUSCICICICICIPieWOCCSeZ7Y4o2l73OOA0AZJKirRvGm1X/iHPpk07ooJ3f8Ah1STjvuXwkHGDyJ/zAKYjYlhFAIgIgIgIgIgIgIgIgKJe0Hq72O0zaTo+VVWU7ZZpdwwyLeMsx5vAcFekblE+ka27REt60NeLyYZI5qBrZo5XvIJY0HeB6Y5/wCQLM4EavFj1ZHZayU9xUM7t5zyycFj/wBfotp/FEwiN/KzAIIBBBB5ghF5lhEBEBEBEGo1VqG2aatguV1nEVN3ndlwBJJ2uOAB1PuqpXE/iNqfiBqB1BQGqpbcZ9lLSRk8/AF+Dgk/zIWuKPO0S7rhzwGZV0slwvVXUxSuqQBGY9m+LDTvB8HZyMf3VOOnNF6bsNE6loLZBh5y+WRgdI44xzdjPRL5J9GnQoskuc1RorT+oKB1LWUELSakVQkawbhJuaXO/wA2wA+ir/xu4X3Gwm66qp5IJbW2VsoY0ubJCHyABpc4nIHic+K0pf4RMbdx2ZtRVk9DPYK6aWoEIa+mc55OyPGR18+fTyU1qL+yJEVEiICICICIIz7QmuodI6Qmo4pQ243KCSOn828sZ/NVEpprlDVwXGqmm3PjIhk5j3OYOD5ZJ+63xxqNolbbs5awi1DouO2zPebhbnPhkDsnLWuwDnx6qUVlaNSkRVBEBEBEEYdpG8fhuhW0jXTNdWTbSWD3drRkhx+ZaceOFW7Seo7hpq/R3agqn08hGSWk+8wkHaQeRHIcltTXb5R8rnaVu0N905b7vA5jm1cDZDs6BxHvD6HI+i2SxSIgIgIg86meGmgfPUSNiiYNznuOAAqfcbNWP1HrOqeysfVW+ObbSMB9xke0Dl/ixu+qvSPKLek59na8OrtO1lHI+N3cz7mFv94cx8uWf8ylJRaNSV9CKqREBEBEBEHAcetSUdg4aXvfWdzWS0vd07A7DnuedoHy659AVT7S1VDb7ubv7XNG2mZH3eY+b3kDfjrjBJ+yvXalt+oXp0beqTUOmqG7UcjpI5omklwIcHY5gg+OVt1RcRARARARARARARARB+SPbGxz3uDWtGST4Kpmrb7DrHWt7uAxF7SyOlgw7c6JjHFucev6rXFHnaJWa0/aaFui6Szud7ZRSUQieXE4lY9vveRwcn7qpGpLVJpHXU9BteXW+fumyTe6+VmMsf15ggqcc+ZRK13DW6Mu+ibZWBzTI6BveNBztK6NZWjUrQIoBEBEBYd6udDZrZPcrjUNgpoGF73uPgBnl5nkkRsU61Lqa/8AELUc0UksroW1T3U9NEARGzJdyHiVYDgrw3odPWymu1dE+S4vy8CaPDojkgO+Zb+q2v8AhhCUUWKREBcVxvtNzvvDS5Wi00pqqipfADGHYJYJmOdj1wCpj2NDwG0LctK0ktZdWtiqJWOhbCTlzWBwIyfopTU3mJnwiBFVIiAiAiAhIaCSQAOZJQU54u3mr4jcW/ZbbGyWngd7DRPYw5ewPyT65OVLvE7g5R13C+ht9hpz+L2eEOgcPiqNsZa5h+f6gLaZ7dCEeHWtrpZbtZpW1De4hq3CJjowNpe4B7Hu+Lq7xPL3ldKkqIKunZUU0jZInjLXDoVGT7oh6oskiICICIIG7XFwAobNamvIcTJO5vgejW//ALqH4NP1Ffoqa900J3W6Ta9o8Yz1P0OPpla19K/KUOzlxB9lbT6Wr5CKUuIpuW7Y5zumevUqxKpePKazEwIqpEQEQQ92pdTutekYNPU8kQmuziJRu98RMLTyA5jccDPzCrLFTGoo66rOww03ds8sF+cY+y1rHhEpy7LFxI1BcKBoaI5KZr42D91reTvzx91YlUv7IEVUiICICIC+ZHtjjdI9wa1oJcT4AIKWdpHWcmo+KVbSQuY+gtLRTU23nvOA57v9Rx9F+2XQVfX8LblqKSjdAaKP25neRc5Y8DLR5eY+q2+ETKW+yRqaSvs1ZZaieLMGHwxl/vY9B5f7KeVlb2QIoSIgIgIgIgIgIgIgi3tJX2S2aOo7ZT1EsE10qxEZI3bS2JjS5xz8wwfVRBwesDdQa5gpn0sklCd9RPho9zqOvgDn7reniu1Zja2Kg7tRaejfS0epIY5WTRgwyyxjOPFmeXLx5rOk6smfTx7NGpImSVFhq6tzpJIw+AO6e5u3DPyx+andMkTFvJWdwIqJEQEQFXXtLapluFyi09b6yKWjjj3P7qQHE+XsIPqOXL/dXxx5RPpsey/o2jNkGo6rZLUNqXBgLCHMc0f/AOsqeUyT+JMCKgIgIgIgIgIgIgIgKPOO2tKTS2kZqMVPd3K5MMNM0dQDyc/6AlWpG5EU9lzTlTdr9PqStcWx0MbYYsBvvO3l3l57uasypyT5RCmvaA0nHo3iJLJTd37LeJnV0LQTmM7muI9PfB+4U+dnXVJ1BosRVNXDJVQTSsbE13vtjaQMkeAyeSvad0RCTkWKwiAiAiCuHa2qmHVFjpHSbWR0cksgA5+8/A/QrpezhaKeq0zdhVNEkNUxkboy3ltLSDj81p/CrryhmWK5cPeI8lNGI3z26f8AZPlbua8D4H/PCthw+1B/SbS9PdS1gc8lrtvQkJaPG0UiY8OgRZriIC86meGmp5KiokbFFG0ue9xwGgdSUFJeMOqn6r19dLvG0tpw4QUgDST3UfJpPqTl31wpEqeHN1tfZ/v1TVRUrKurMNYxpLt8IY9vP5lu76PK19K6203Z1uFLScYIGzxNp21cEkUDWu91ryMjPmTgj7K2areNSmBFRIiAiAiAuH466odpHhjdbtDIWVRa2Cnx1L3uA5fTcfokIn0p1w90/U6r1LSUsPemorqppklfzwDzJPqr2w2agp7EbNT00UdL7P7OGbRgt27efmr2VpGlSeHNVUcOOM7aKvA7kSCnkcx3Jof/AM/VXFY5r2NewhzXDII8Ql9e4Wh+oqJEQEQEQEQEQEQEcQ0EkgAcyT4IKr6q1L/TnWMVXVTSstcE0ncsD8/s+eMep9wqVuzlZ3W/Tt0rJo3Mlqa7DefIxhjS04+b3fkt7xqulY9pSWt1VRG5aYutuA3Gqo5YQMZ5uYQP1WELKg2aqn0Vr0U3fEuiJjmwOfInl/c8Vce1VUddbKWsilbKyeFsjXt6OyM5WuRH6slFkkRARBo9e3lth0jcrl3rY5Y4HCAuPIykYYPvhVe4cWWTWWq4aB74cSRk974AAEkkfxcvzW2PxEyiVr9P2umstlpLXSNAipomxg4+LAAyfU4WcsZ8pEQEQEQEQEQEQEQEQYl6uVHZ7TVXSvl7qlpYzJK/GcAKovGHUQ17xM30s7vYqeljihYQQA7q/B8feOPDoFpjj5RK02gLLRWTS1BT0kMLHOp43SujHJ7to5rfqkzudkRpFvaV0q3UGhYrlG0uqbJUtrGtDS7fH8MjcD0Of8qhXgfrSOxatgfUvghpJ6lwqpcYaC7k48uoyT/pWtPNZJW7aQ5oc0ggjII8V+rFIiAiAiCpnaPrBV8XKqmJLo4ooIuY6Hbk4+6mbs4QlmgXSuGHPqXNI+TR/utJ/Kj5cj2p9HD8NpNW2+PD6eQx1YaObg92Q8/I5H+ZfnZk1lvpY9MzTtfl73xNwMtHzHqkeanynxFmkRAUP9qPVQtGiX2GKRjZ7qGtJa/D2xteC7l5Oxt+pUx7RKC+B+lp9Va5pKfu2PgpnsqqjLuQjY8ffngY9Va3iwGnhrqBrwS19DI0gdeYx/NWt70lUrhtKYuIdmrIab3YqqJwYX46PBwruJeNIj7CKiREBEBEBVD7U2qqq76un05HVj2ekqIyyNr8tJ2Yz8/fKtWfKJd92TNNBlDW6hqGU8rHbIqXLPfjwMl2fUFv2U/KLT5IVZ7X9nmt2s7RqCliMVPVU5ZLI1uAJmOzuJ88EfZTJ2f9STak4d0k1QxjZab9g4teXF2B1Of9ynwb86SEihIiAiAiAiAiAiAo748apFg0qKCJ0raq5h8LCzltbt9458OoVqRu0CCeFlsmr9Q2e2u7x8c1RvfyPw48vEDr5clbG2UUFuoIqKmBEUYIbn1OT+ZV8s/CIZKLJKqfaUomU3EO5VbZnRmaGBzIyzAkdgbsO+g+6mXs/Xxt20RDE6Zj5IABgDGBtHu+uOXP1C2n8iPlI6LFIiAiCv3ai1BPHfbZY4XDuRTOkkOAdrnH19GfTK7LgNpymoLULwYtlXNAIjgjoTuPr4N6rafFFfO0nosVhEHLai19p2yGeOpnllmgcGviiZl2SMjrgfmuSr+L5E80VBYXuEeMSSzgZHPngePLplcPl9cxYdxTzMOlg6ZlyR3T4hqXcU9UzRxGOjt8TpHO+FpOAPmef/dfI4raniqpTNQ0b6eJwG1rDvf+a5MftHkm2tw9s9IiKzLr7BxW03cJ20lcZrbU7SXCZmWcv7zc4Hzwu4oqylrqdtRR1Ec8TuYex2QV9JxOfi5Mfhnz9nIz8a+GfPp7ovc84iAiAiCHO1Fq11s0i/TNFj224taZHH9yHJ6c/iLmgfLPouN7L2mPb79LqKup4pmwwubmRjXDe/GCOXXb/wA5raI1TaPlZRjWsY1jRhrRgD0X6sUjgHAtcAQeRB8VTvjRp+r0XxAn7iOKOhlrTcbfimGwZeHmPb0LGnIx8lpjnyiVmOFmr49ZaXguBiMVU1jRUN/d3c+Y9CBn0zjmusVJjU6SIoBEBEFKOKFRV13EO6y1O3vHzyY93mBvIH5YP1VquDn/AOM7IN0bg2FzQWDAID3AZ/vY6+uVpf0iHS3KjguFvqKGqjbJBURujka4ZBBGCqb6fluvD3iVTWysEffUFaI97m7RIzIGS7xGOainzBK41orobnbKe4U5zFOwPaspUSIg8a+rp6Gjlq6qURQRN3PeegCpNxf1JU6r17c7g5p7rvO5pY+uyNnIfU9fqr0j5Fh+zXpMWTSZu9TA1lTX/wBkSPeEXXr/AHjz+gXZ8U44pOHd9Ew3Rije9w9G8/5KJndhUSju1HQ6zp62gozSU/t3uQl+TGzfy59eQ8cq77HNewPY4Oa4ZBByCEtv5RET8v1FVIiAiAiDheNuqYdLaFrLg8s3MkiaGSO2CTc7oD6YyfRUyscT9SX2oulwkghjqKoGTMgGS955MB54+imNfKJXl4cWX8A0hRW9ztz2ty88up+S6JRJHpy/FbTMOrdAXeyPhbJPLTPdSkjm2YAmMj/MAq5dknV9XbtVDTM8bvZa58h+H3myY8cnkPdwpFtUUJEQEQEQEQEQEQFVbjJf26p4hl9ITPS0J9ngbvwHfxO+WVpjjypfu8RVKfAnSsNNSQ6pEu72qGSNsRH9mRI5uWnxBAUrKLzuV4EVBD3aisMtXpD8bhliBo8MkjMG58gLuWHdW4ySok7OmrxYtSiKV0rmVDBF3bn4GP8Atj81vXzXSszrytxTTNngZMwENe3cMjnheiwTE7ERIvKrnjpaSaqmOI4WOkefIAZKComt75JqviPV1ccEbw8kmPdlpjYMAdPixhWl0JbzbNJW2kcI97YGl3d9MkLbJ4rEIjbdosUtLqnU1r07TMkrp295I7bHCHDe8+gUEa64gXu9e1tonVEEG/u3RNfyA8B7vMr5/q/Oiv8As1n+bo8HjWtMZNbiGNZNLV9f7ZTPgknmDO8kbGP7Pl4n7qSrbwsZIIpa+tMbmt5Mjbuxn1PjyXH4fSrc78VvFXV5PUowR2xG5b+m4dWOnoxTRSVOB4uLTn58lq7zw3a7cbXMxnjiVudwOcjP2Xpzfspj1NsV53+ry4etXrOrRGnBam0hU0dxIlg7ve082DwXP26svemZaaa110lOSw42AYI5nocjx8lwK3zcPL2W8TDsW+ny8c2iPaROGvE4vAtWoJJJJu8wypd6no75KXWOa9u5rgR5hfcdJ5/71i7bfmj+/wCr5fmcacF9fD9RdZ4xEBfM0jIYnyyODWMaXOJ8AOpQU013d59a8TrjURtLo9/dsDeZIZyYAPl/NWq4cadg0zpOit0bAJu6YZ3bS0vftAyQSeYAA6+C3yeKxCtZmZ8ujRYLCiLtS6abetDU9zZlk9sqmufIB/5L+TwfTOw59Fan5oHBdmLVYt+trnp+pfFDRV0Ambuk5QzMIAYM+BBd9R6qzSnJHlECKiREBedSXCnkLXiNwYcOPRpx1QUg1BVvut/rKuYs70nD3xjGfAkBXG4fU/suh7LDsjYRRxuIZ8OXNB5fdaXREN6q49reyGnvVn1PDGz9pH7NIR8WWO3A/Y+f7qrX2mXY9nbWkNy07FYazZDU05d3JMme8aTux8+al1LRqUQIqpQ72n9YRWrRsmnqWR4uNxeyMbXY2R9XuPpjA+voog4EaMj1brARXGOV1BTtM9S4Z3PePgBPz/T1V/NY2RPlbyjgZS0kNNHyZDG2NvLwAwFgattwu2l7pbDu/rVJJENvXJacfmqCkl3h2X2Vkjdknf8AdPD8ZYehz4clP9fxeltNhtNvtlqeHNoYwZ6gt2uIaG+5gnlyPvH7Lz8rk/SrM6Xx45t4c9U8ZNZTUQjhmoYJdm7vY6beT9+X5LFZxY10xu515DwGknfSQj9GL5jP1jkzOqTr+kOrx+FjtEd7Xv4kax9rMovNd3j2khgf7jfpjHisb+nGp5JKd8t4uMkoBwfaXjr9fJcy/UOTM/nl7o4+Cv8AAyaDXN/p3skbea7cX5xLN3nh69V7/wDVDVrXh34t7plBb7ke/PTwGMc/JZx1DkxvUrTxsNvdSTiRq2aXeb5Uj9pn3JgAP7uAF4R8RNYOkc5+oq+EF/ugPYR19R81enVObv8APP8AVjfh4I/hcNxJ1petQSS0N5rJKiOSQB8WQwHHozH/AALX6fmZQd3UUlMwGCRj8n3yCDkcj16L3RzeT27iXkjBj35hIEHFPXUNRUd3fKnu388d1E7HqMsWTFxe1y3u9t4llxHtIdBEMu/0fzWM9U5cR4u2tw8fvT1dxd1wHNY2ueNoI2mKP3vrhRRX1jqPUjrqWPhqpZjUZiIA35z0xhb4upcuZ8ztnbj4onxCSbLxX1MylhNPM+DmMGPofDmw5Gengs2TijqtoL33StyHe/7zB18MAfPovNl6hzY122b4uPx5jzV6nivqORjnfiNUCXjbl428jnyWxPFrVrtrBKM7HOwGjwx448ua83+q8+v8c/1b/uvHtMao86fi3qsND5K7eR8Q7tmOX0WeeNOpC5jmMoWMbjc18Jdu+uQtKdc51f4t/wBNqW6fgmdaedHxl1NJWwPkqKUw+0hr42UvJ7PnnIHqsyn406gMA7yK2FzZME928Ejy64C9EftFy6x5iFP9LxW9NhLxluntEwFLbYY2MyNwe7n9wtppri1UXGF0Rt8E87OW6N5AP0wr4/2n5Ez5xxP8lLdJxxH5ktIvt3BEQcrxQ1HS6f0rVmScMrKmnlZSMwcyP245HoCMg81XTRVjjulzpadkVU+aqBDDgEh+zzHLGeeThb441Eyid/C1FjttPaLRS2ylGIqeMMHqfE/MnJ+qzVhM7SIgwdQWunvVlq7VV57mqiMb8eGfFUfvtvrNH8RK61Vk0Taiik59y/Iw8A56eTx91rin4VtG1y+GdbLcNCWirne18j6du4h2T9fVdGs59pj0IoSLleLlY+g4b3ueMe86n7oHONveODM/TdlTHsVv4Pacqb/xLno6prAwUhmD3dNgZGABjx98c1benijgp44Im7I42BjW+QAwAtMs+dIrvXl9rn9fag/o5YHVrQ0yueI4w7zPivJyMk48Vrx8NcNO+8V+6s1xulwvV3dVPeHjeXZPV5PIYUncNeG1bUQx3K6kUcJxJBGP7Zx83eAGD6r5Ph8L97y/i9fLu580cfH+FL1otdHa4HQ0cWxrjud6nzWavrsOKmGkUpHiHAveb2m0iLRV8TRRzRmOVgexwwQVxt50DS1b2SQThhjDtjHM93muV1PpdOdWN+Jj5e3h8yeNM/MSgvVtjuFoqS7+zmp5zmM9MDqf9uakfg7rwSTutV1mY0SvAhcT8L9o5H5/86r5np154nJjujWvEuzzsdc+Hvp/NMSL7t8yIgLg+PWo/wCjnDG7Twljqupi9mgjcRz3kNccHwDXE/ZTWNySiLsp6aFVcq+7VNut81O0hg3jL4sh/Nv1936KzKtknyiNfAiokXxPFFPE6GeNksbhhzHtBB+YKCkusqabTPGu7Qso4mvo64TRsIw0s3iWMcvDZhXUtVYy42ylr4mOZHUwtlYHYyA4Ajp6Fa39RKGSiySIgLXane6PTV0ezk5tHMR89hSBR2IS1FbI2Ml0kmIyfHPTkrz2OA0tloaZ0YjMVNGwsH7uGgYWl0fLMWg1/p2n1PpiqtssUckuxzqff0Em07c+mVSJ0lVPhtWs05rulZc55KeOCsY2QMIG3Bwf0VymuDmhzSCCMgjxVroifD9WJerlR2e1VN0uEwhpaaMySvPg0KiVJdbX+5cQOIEt4LZGsqZBFSw7c93EPgH+L/dWt4OaQ/onpwxzwCKsqSHTDl/L5rS3iNGncIRkEZI+SzFIOKdPT27iHfaeiZMKWG4P7sSfGHgjOT5F+SPRT9orQln1Xw4tnthmiqYYXRNlif7pPqMnPP5fzWOfj1y+2uLPNPSLtV6Su2nKx8NVTERtHx45P6dD49QvjTlLba+pEFTU+ybx3eTHnGRyXwnI4l65ZpaPl9Bjz1nH3Qluk4K2zEL33OZobHjDGLa0nB3SMLWd62qme0Y3GTGV9Bi/Z/DEfjmZcy/Uskz+Hw9WcINGNj2ez1RGc/25XpS8I9CwPY/8JMpa4O/aSEgkL24+j8WnqrCedmn5e44XaO2uDre9xMjpAe9ILc+AxjkFy+v9AaL0xpe6agnjmENNCXNjfLy3kgNGeuMkevqrz0rjT/Cr++ZfuqBH3131HUHvoxEakM98ksGeQPTp0yrScK+DlBUaa77UEUrJpXNkjdEWgFuOWOXTH6qb9NxWjUwrHJyRPl1zODemmOOKirLCMYIZn74WVHwk0nHUtlbDK5oGCxxbz+rQCvN/ofGbTz8svWDhRpCMkupZnkjHOTCjvtBcKaVmhoq7TNqjkrKGSSeomMpbII9hLiB0IGM49B1Xpx9M4+OdxDPJy8t/co27M1FZ9Q3F1lv9qhqnPf3sU7a0xvwMgxlo5kc85G0qwNNwZ0fFSVVM+OqmE0m9jnTHdE3wYD4jp158uq1ng4Z9wyrnvrxL3Zwf0OymlhFtcS94LXumeXMHkOfz+6+Z+DuipJXPFHPGHNwGtqHgB3n1Wdum8e0eYaRyssepYtVwU0jM4PYayI4w5okyw/fn+axKvgbpyargey4V8cAGJ4gW+/5bTjlzxy5rG3R+PPqGsc/NEe3Max4daK0vDSd/frjHUBx3RQNYXy9SOvweWVG9fBRs73u2PEW/DBIRkjnjPmV8f1jHhwZvp4p3Pz/0dnhXyXx91o/k99OaduOprn+HUMTnSSxjLtp9weZKsJw44bWzSjBUSvNZWkc3OHuN+Q8+vP1K6PQOn/Vt9a8eI9frP/08nUeRNf8Abifbu0X2jiCIK6cebz+K8QnWbvWGC307D3f/APJ8W7Ph1I+i7fgvpyEmHUkYZDD3PdwwsbjmWjJPpjp81vM6ppXXnaVEWCwiAqodrOwexa7kvNC2IOrKOGaYEYJkBMefXkxivj9on0kTsr3iaq04+2yd4WMZ3rS9+fEDp4Ka0vGpIEVEii/tJ1NTHoqjo6cODauvayRwdgBrWPdz/wBOfmArU/NCJ9OH7NTW/wDUK6whzHmltoYCM8svb69OQ8FYhTk9kCgHtF6m9ouUFlgZujpne+R8XeHI+y5fU76w9v3eviR/uxP2b7gvoGFlFSahusbXSubugi28h15nKmBX6fgjFij7yjlZpy3/AJNdqC80FjtVRca6ZrIoAMjPMk9G/MqKdUdoPTdufV09so6itmheWRyOGyN+M5PPBxyXRrTbzImvvaS1rcKfNrprfbY25aXgd49/8J5/njzX7aO0ZrBnKvq6SUGXBzTsa4N/xAY/I9Fp2VViUr8PuPtkv12jtl5p4bXI9hPfGX9mCAMAgjoT0OfEKZKWogqqdlRTTRzwyDLJI3BzXDzBHIrK9e2VnL8StPRXu0wvEQdPTzscMNyXNJAI/Q/5VXfUNkmtdxnc98ke6c7yMjB8cL4/reOcXIi8R4tH94d3psxfFNZ9wsPwn1JHqPR1BK959thgZFVNcefeNADnfIkFdavpuJmjPhrePlx82Ocd5qIvQyFVrtIahobzrW5WVlNJLWW6KGlge6T3Iyf2kkjR582M5+R+mmL8yJnSZeCWmjYrDJPKYTJUYDTETt2jy9P+6kBVvPkgRVSIgrn2rtFbJX62pjFDEY4oatxfh75C9rGbR8uvyXX9mrV0V4067T7n5mt0TDGCefd9Prz/AFW0z3UV87S6ixWEQFzPFebuOGepJCSP/DZ25BwebCP5qY9ipeg6cVuoaKkdFu7yqjY8s5u5nGB65Ku0rXRAiolWLtJaTjsuraS+0ndwU1xe8ujH/qtwXHHrn9VLXAC+yXvQobPK+SajmMLnOOeW0EfqtJ81QkNQL2ttU91aqPSVHOzvKl3f1rcc2xjGwf5jn7KtY3KXNdmLRL67UYv9WxrqS3YewPb8UpzsI/w8z/pVn0vPkEVRTLjnEyn4mX8EuLzV7zuHL3gCPpzVkOAAe3hZa2TPifO3vBLs6g7zjd64x9MK9ojW4REQ7O626iulI6lr6aOeI+Dm5wfMeRUIcQ+HEtlnkudsjmqqXZl2ORj8P+FcPrHEtkxfUx+4/wAOjwM1a37L+pfeiuJ1ZaBDQXKF9TTZ2jLv2kY9M/EpjsF+tV8pmzW6rZLkZLM++35hU6R1GM1fpXn8Uf3X5/D+nPfT1/hs0XccwVdu2tqh1Npy36Rgad9c8VNS/ONsbTtaPqc/YKYRKKOzxp2K5a1tcdVTONLLVYeXMcWPLBvLPqAefRXgSSK6EUJF8VMMVRTyU87A+KVhY9p6OaRghBQukhruGfGEUc1LNCbdXHLc7d8RO5mD06EH1wr22mtiuNspq6EgxzxNkBHqFaUzGmUiqh41tXTUUBnq52Qxjq5xwFGPEPipT26I0tmy+ctJ73bnl6BcfqvU6cSnZWfxz/b9Xt4nFnNbc/lQZcb1cL7XtfK+aodLN+5kvJ8OX/PzUg6C4S3q8SfiWoKiakpJOYhe494R8vBfL9P6dPMy7t/OZdbk8qMNNR/ROWntPWewQmK1UMVPuAD3ge8/HTJW0X3mLFTDSKUjUQ+eveb27p9iLRUWm1teWaf0rcbu57Gup4HOj3dC/Hu/mpiNyKx2Kkumr9b1lRJWmulr8DvANrZBnlhpHu4z+qtbbaSKgt9PRQNDYoImxsAGBgDC0yfEIhkIskiICjTtJ2uGs4U3WsZRU81bSiJ8UrwA+MCVu7DuvQnkpr7J38IS4GX2v0lrtluuNKYmbB37N+Qxj8e/yODyIVt2Oa9jXsIc1wyCPEK+SFa715fqLNYUYdo8VDdG0tRE0mOGrDpCGZxlpaPlzIVqfmglxnZka1mr7uGwxF7qKPviOsZB5fcEfZWCU5PaIeVZOylpJqmT4Io3Pd8gMqoN07m5XzEUT3nvjkv5k5J5/wA1wur2mOyI/V0eBji0zMrb2OMxWajhPWKFsZ+bRj+S+NQXWlslnqLpWl/cwAZDBlziSGtaB5kkD6rr4vNKvBb3KnXGHW931TqC6QQiqoreydr5o5m84S1mxoe3wIyevmtlwq4LXjVlBU1tZWSW8bzF30jN7ngjJ6+ORg/NeudVjyqmbTfZ90LaqJkc8dTXT5y+SV/InOeTfBe994CaBuscgdRy08hwY3wu292fHACy+pJMbQ3xE4E6m0rUV120vVmstgpg2QyY7zGW5G3PoOY8li8D+KV10zUU1huFS99BHNhkcj8BjSfh/NaRq0C3dsrqa5UENdRyiWnmYHscPEKNuKGlC4VNxYGmCWTJYOuSP98/dfPdd4ts/G3T3Xy9vByxjyefnwi/Qd6dpXWguD95Z3b4pIyeoz/2VmrfVwV9DDW0z98MzA9jvMFeb9n+RNqWxTHrzH67enqmGK2i8fL3RfRuU4fjlqgaS4c19zbnv5HMp6fDsZkceXP6FVz4RabuGu9aXK4up6Yf1zfPlxd3bd3vOZuPmcfvLak6jaJW/pKeGkpmU1PGI4oxhrQegXqsZnaREBEHL8V9NjVvD28WFsTZZqiDMAccYlaQ5hz4e80c1WjhJqGfRfGJsM1DFDT1ERoKqP4GQc2EvAHjlg6+ZWtPNZhW06W/RZLCIC4Xj7M6HhFf9rXOMsLIuQzjfI1ufzUx7EHdn+yMumt43Op5Hx0wE7nBxbsweR+/6K1atf3pECKiXE8ZNGN1rpZtGzaKqll9ogJz1DSC365VeuBl8fZeIdHSVYlZDJP3MwJI2PPQkfPktIn8OkfK196uVJZ7TU3OulEVNTRmSRx8AFSnUd0r9e60ud3iZPuq6kimjkdvLGdI2KKErl6Rs1tslipqO2UzYYu7aT7uC44HM+q26rKRFAqb2nIIY+JNzqpHYl9mp3RjHXDP+xU29nRpHC6ieS5xfI8lx/eIwHYHgAQRj0V5n8OjSRV8TRRzROilYHscMEHxVJjYj/X3DSgvkza23MjpKhrCCG8mnlywByyokjF80rdaTvDNSVLG5cScHoTzC+N6xxL8TJGbF6mf+Uu/0/NGak4r/wDcJe0JxGo7sI6K7SRQVruTXtPuSdPseakBfQ9N50czD3fMe3L5nGnj5O34+AkAEkgAcySqI8Y7rJrXihcqmkAkZ7SYYZBzDowcM/ILpQ8czpYnsw2KKg09VyTW+ISQ1P7Ccx8wSwNeG+XJrf8AV6qZUn2nexFAIgq521bI6iutm1TTwt21LTSzOazmXtBLdx8yDy/wFSX2WdS1GoeGrI6iFzDb5fZmvJBDmho6fmp+BKtRNFTwPnme2OONpc9zjyAHiuH1pxHttlmNJSAVc/c95lnMNycBczqfUK8LF3e7T6j7vTxuPOe+o9IV1brS86groTKXvxyYGDl9AF66f4e6q1PUQyPt1RR0xYHe01I7vlz5AHmvjuLxs/UM83nzM+5dvJmx8akViE06F4a6f0zFBP7LHPcGN96YkkZ/ug9F26+54nFpxsfZVwM2act+6RF6mQiAq7dqbWEzKqLTlM/MLHME7Wkc3Owf0KvjjyiXM8NdWt0cH1cVJHWzsg/80Oaef+HPqMrsD2kqZsLw7TDjMAdhFZycfDlsz5LS1O7yb08oe0zTmeMS6QlZHJGHACvBc0+OTsx5f8PLLf2kbcJaeJmmahxmaMPdWMazd5ZAP/MKv0pNtcztNMLY3v0oW7HvE8ftgzj9zacffkvWt7TMAD20ulH5Y8DfJXNIx8g3OfRPpI2/aztKNFFNUUukxmOYRhktyaHOG0nONvIZGM81jXjjjR6nst6tVx04YKWppGspgK5jy2TLiXkgcsEMI/wpGPSdoSrfa7VqWRs0RgnIY/nIP7MgEEfQq4vBvUsuotGUL6tgbVxwgPLTua9oJa0588AZHmVN6+NojxLtUWKwoz7Sznx8MJpWse8R1UTnbXYIHPn6q1fcIlwvZbZIzU1zcWSt3UDXS7vA7m7Mjzxn7KwynJ7IazVh26Wup8qOU/8AsKq1pylbDqOJsu8B5kGXgvOPIDHoF891idXpDpcGfEwtlQtayiha1+8BgG7z5dVGXahu81p4Xkwz9w6qroYd/LljdIPzjC72LXjTnz7QvwMsT9SVrrcYXSsmqfablM7rkYxkZ583q19yraG0UUlZVvZBCHZc7GMk/wA1fLeI9/ClfL8tF1oLtAZ6CoEzGnDsAjafI5Was4mJjcLTGnxUQxVED4J42yRPaWvY4ZBHkqado/QFRo7U8FdbnSyUdax0scgb8D2/Gz/4/wCr0WlJ86RKSeyVrt1ZSyaQuDnuqBuqKZ5/hwNzf5/QqfrhAypoZqeQZa9hBVM1dxMT8rVnU7Va4g0kjb7IWhsMTIxy6ZJyp24J1wrNB0rBu/q7jHknOfH+a+T6NqvJ1H2l2+obtgiZdsi+scNWXtf6rhnvFr0nTvMopm+11DY5iMSHc1gI6ZADjz/iXQcEqrTXDnSTai9T9zW3EAsEbe93MBPTaOXM+PXAW0RuuoQkUcUdGOilmhubp4oXsZM+OMkRl+cbh1HwlZP/AFJ0J3xh/pPb+8EfeFu85x9uvp19FTssRO2TR660hVtjMWorcHSQmoZG+YMe6MDO/acHGOfRetRrTSNNRQVtRqW1RU9QHmB76pgEoacO28/ewfJR2ylmUl/sdUKQ093opPbc+ygTtzNjOdgzl3Q9PJesl3tUYiL7jSt76buIsyj35M42jzOfBRqUbh9SXO2xslfJcKRjYX7JS6ZoDHfwu58j6FVE7SVELZxZqZ6SjAgkjil3tePfeWDLvQrTHvZZaDhdfm6l0DaLxv3yTQAS5IJa9vJwOPHIXTLOY1JHoRQkUX9p+slpeFU8UXI1VXDCXZxtGS/89mPqpj2iXDdlmjqHagrKx0bdsdLh7snq5368v/krEq1/ZAiokVUuOWmZ7FxGqqymp3R0ta/2iGUMDWB55kD1zlXp7RLz4n8Sq7U+gKHSneTCrdIx9bKCNssbRyZnqTnBP0XVdmrQkUlRLf7jTvMVOQKZruQc/wAz/FhW12wan5WHRZJEQVQ7UXcP4kVjACHClp9xx1dg/wAsKV+y/cI6nh+aJsTmSUsg717nZMjnZ97yHJoH0KtrwjaWEVUi0OqtJ2bUoYblA4yRghj2OwRlZZ8NM9Jpf1K+PJOO3dVBustBXmyVsckLMxOkcGSM5DaPhz69F86O4pXuwVVJQV8prKMvDXtlI3sHofsvjcf1em8j1/8AjuZLY+Vh38thx74uwu0ZTW/TchZPcy5kzycPhY3GcjqOv5KDeH1iqbnfbbHHE90dRWx0pk34B5jPP6j7r7Lj8imakWr8uHkxTX2vPZrXQWei9kt1M2nh3F5Y1xI3HqclZq1UEQEQcFx80vHqzhncaIgmopsVdLyzmVmcD6glv1VUuAerJdKahirZXvhg7zu5gxnUbzkEH6rPLljFSZa4q906S9rzio+8UNVT0r3sonY2R+IAPifH5FaCxWDVWrKqYx0la5nJvelhAZnzJ++F8Rmxcjn8jumN/b9IdjDkpx8eolN2heGNj08IquoY+tr24cHzHlGfQDku9X2PD4lOLjilXHzZZy27pEXqZCICIMS83Kis9snuVwmENNA3dI8+AVPqAVGtdbSXGoia/vJi5xazu97AcB/Tp5fRa44RMpfuXBq8VTAIrxSR7W7GlpcwuBHMu90n6fmtG7s+X3u5YxqO3vYWnYHwv6nGfpy9fzU/Vg0wZuzjfXyB341agADhu1+PQfD+awKrs16qc+FzL5aD7+ZQHSjA6nBxz6Dqn1INNdX9nDW8bmMgrbbUZkBke2ctHTr7wyvyv7NusPaJDBWUbgfHvSW/F/q6c/sp76omJe1X2bNZNgeY7nbJQ7ADGyO3O8cuy0BvPyJWDL2euIZO6H8Pha1oY0moy793ngeHI9D5KO+qdNNxB4W6w0xTfid8/awd2yPvmTd4GZIGDn1wPqpO7J2oZ21FTY6p73ROhD4yWHrv5ZzzGdxVp1NfAsWi86RR52i6B1dwmuj2Dc6kdFUhmcB4a8bgfTaSrV9wIn7M9aWavq3RU1aWyQME7mDLG+Ddx+mB54Ks2rZPaIliXuH2mzV1Pt3d7TyMxt3Zy0jp4qqd0L7PqCWpqWZkEhaRk8z0OF871uJ/BMfr/wDDocHfmIWJ4VX2C76Xpoe/c+pgZteHuy8t8Cf0+iwOPumXar4YXKghgknqYCyrp44/ie+M52j5jcPqutwcsZcNLw8eWvbeYQFwQ4ht0TqA2+8U+2OeqfTVocD3lKc56Yz7pBBCnrjTQuvfDmevoK17IYIfay5hwJIgN+efyBW3KrNqTEfZXHOrOL7Mt9qpq2uslQHvb3HfteWHwc0Yz0/eP2U6LzcGZnDG23JrFckxAuL4waNGsdLuhpmRG6Um6WgdJ8IkLcEHPLmOWT0XsrOp2wlVTgbe22nWdkqJHOYfbxEdrNuY3kA5Oefy9Arvq+T2je1feMFEy3U87GYHcZyT75IJ6deuDnn5LqOznWOltNTTF+Y42jYCeYOTu9PJfEcGPpc+I/WYd3kT38baWlq9V3hlh0/V3V7Gydw3LWF+3c4kADP1X2kRudOHKglfUXC8T1VyYJ6iqE3eGXeSefQDmcH6/plZVPYNZ11vrKgR18xpp46Yxl73ljywnH2/VeiYV28I9LatMdR3dtqwWTDf8bzy8APmPyX3cdJaviqG+00NcZO471/uHLM+eOnXxUaH1b7DdZO5ZWUlU53dvDO9B9z3MAgnqAfALxr7FcRTUhNDWvmwRIwkHeCfjHLkPv0KtpHnXh+Chre4p6jvZ5JA88mHBHMYI8uo6ea/GU7nz0tPU/iNF3tTskLDsyd+Rs8jz+pUSp22iWquktfW1la+qdX108uHvkm3veSORL85OceOV81F0u0shpqn2ieTu2YM2SQGAYOfQABRuY9NFgOxtqiop7tLpdzHuo62KSojcfCVmM/dv6K0yyv7TAiokUSdquoEfDWKnYWGaevj2Anm0Na4l2PLoP8AMPkZj2iZ1DV9k8N/Cr05zsSufBhp67MPwfuXD6Kb1N/ZWdwIqpFxfGHR79ZaUbRU7j7XSztqYG79okIyCwnwy0kZUxOpEF6F4WamvN3gfd7LUW+mhq2tqXSyMG1o5lrQHEnw5/8A0rQ0FHS0FIylo4GQQMGGsYMAK15RD3RUSIgrP2pYtup56qOFm9lHD7waSXc35zg/4V0XZGro32q80DXFzmPil+EDbncCPv8AoVpP5UfKdkWaREHxNFFPGY5o2yMPVrhkKHOLXCuhq4oq6z0/dz9+A2NvP4gftzXj5nFjPTWvLXFlnHKrF0qRNcJqCRz90Ej4snwI5HxXrRXe8WPu6mm7wiKQPD4345/yPLqvBx7fR8TLbJE2bSl438RKYTj8Tq4wZ+8wCGEuznBJz8XyX1Dx917NC6GpuBfD3u7JA3/cAZHzXax3i8bh470mGzb2i9fGlnY2pYXFoLX7Mcs9M8/4v+YXwe0drptZJUR1MIdJS+zjfT/DtPxdeZ688eJVkRW0uip+1Bqplufi30UsracAucwkghvOQdPng56+ixB2oNWPyAIQ50Ja0GlBG7+MEfkFFpivmURMzLaUvaL1rVUMLWUDGzEkmoMLQz4cYx8zn7KKu9pGV73gSVFRUTE4YOe88+g69Vx+RltkmIh7MUdsJx7PehLdql1Zc7vUVP8AUpYw2nY/Zzy5wz6dVZakpoKSBsFNE2KNvRrQvdxMUY8e/mWeS/dL1RepkIgIgIghXtJ6sig09+D05d3jqlrJMSBvP18cLVdnnRrJnC6VQD4osOLXtw5+c7T69M/VbxHbTbPUWlP6LBoIgIgIg4njdZxeOG90hawumjYyRhGTtDZGOdyHXkCqxaJvTLBqunqI3zsbDhxGzGTkE8/Lr91tj9KWvqdLl2ysguNtpbhTPD4KmFk0bgeRa5oIP2KyFiuLB1Da6e92OttFU57YKuF0Tyw4IBGOSROhT/SldWaP4nSWox1sLaW4Mhnbu5sAk/MkPyMdcq4dpr4blRNqoWyMBJBZI3a5pHUEea1yRvyMtQhxy0u2mqaetoYyW1Hebxj3WnquN1fHFuP3/wDp8vZwZ/3oj7o50dqW56Our6yKdw2xbXxu5ggnpg/85KaNFcWbXdTFBc3RwSS/DIzp4ciPr1XL6d1CcWqWj8P+Hq5XEmfxx7Zd+0jwoqLi/UV1pLC2plkL5anv2s75+f3sO945/NYPETXGlbfoSts9lqqapkmpzSQwwDLI2vbjPlgA/ou/n5eOtPzQ52PFaZ9PPs72RlPY6u9SRyMqKiUwYe3HusxzHzOVKini17cUJ5FptknYi9DFRWCMWbWdVGxkfdUV1eWPYCBmOckfTkr1LXLGpRE7lDHaIjiDQ13J80YLTz6D/hWF2d5+4miog+Rzsuy3OQG4ec/dfF5KTj6lH/uif+brRueN+mk6Kt3bE1XVRXCyaSoSG7QLjP8A33biyJn5SE/Nq+yp+ZypbPsxaDhgskV7raeFwdL30Tjgu3+Ax4bfi+ZU+CKMb8RsG/4uXxfPzTJO5Ifu1vPkOfXl1X4IowQRGwEN2g48PL5Kg+DS0zpO8NPCX4I3Fgzg9R9VixWa0xCIR26lb3RyzEY5HBH6EqdyPCHTVghq4aqK00jJoXbo3iMZa7z+fqvh+ldOulhlNnow+GYTMIiHJ4zg/mVPdI+f6J6cElZM2zUPe1jxJM90IJc4dCs9tqtjaX2ZtvpRCWbNgiGNvko7pFR+J9BDw34ziPTlI63UkIiq6UMfu6gB/U5xnIVt7Bc6a82Siu1I9r4KuBkzC12Rhwz18Ve/mIlETG9M1FmkUCdqyo7mrsrGc3uik3D0yOf6q9PZLoOzNQiDTdfXBhaKiSNmf8IJ/wD3/NS2ov7BFUEQEQEQEQV97V1DWsEVfE1opZY44XkfEXZf/wBl4dkGeH229wB7RM6Jjtg8g8jP5haz+VHysSiySIgKPOP2vXcP9CyXKGmE9VUv9ngBIwHEHngg7vlhBTbhxYKzUWpYKWkBkMj5JCSM9BknCmbW/CS+WCiZWUgZVU872Nd5x5/iGOXzHLmuL1LBeKzekbiP+f8AR7sF67irk9O0Gn6W5S/i+maavghqWNnif+zdnrgH6H7hWNsGiuFGoLeyag0zZpWbRuj7obmZ8CPqsei82Mm8dvfx+q/OwTTVo9NmeFnDsx92dIWvb5d0sas4Q8NZ3Mll0nbmiMfutLRj15r6Bz4mYQ5xN0bwrtVwmFngcJX0pa9lNL+zjyc5J9eXL0UW0enbbJXvittokm2AYf1yfkF8xyufbJlnHjnxDoYcMRXdkkaR4LXe8MhdW99Qse/f+0h90MLeRHh5LmuOvCys0RRUNwpe9lpZpi39g552SMZ8bj5uYD/oK6XE4lq6tZhlyR6h2nZLvjoKr2WeabFc4w7PdDA4M3scfHOAW/VWZXV1p5InYiJEQEQF5Vk7KWjmqpPghjdI75AZKClt8u9XqjWJZM984fVmUjk8My/wVtOHVrqLRpmKkqoe5lDslnl7oH8ltknxpnWmp26NFi0EQEQEQfFRDHUU8kErd0cjCx7fMEYIVM+PtBX2fXFZHJG5kRZnLW8jny9Oi0xyiU6dmTVFRetHNtVXMZZbdG1rXuO7cwk7Wg+TAGj6hS2q3jUpgRVFXO0taKuz64qdSwxytp62SAvkaOW9kbGj/MQzlz/dXb9m/WrL3V3G2VM1TLVOaJcv5tBbkEZ8z1+hW0/kR52mxYt1t9Lc6N9LVxh8bvuD5j1XmyY65KzW0eJXpeaWi0e4QXrXh9Vw1WZYHyRGMftWeY81HlfpespZAWNIaMHfgsz5YXwV8ebi5LUmJ1H2fXVy4eRj8T7Yz9MV9RPTxNcJHGTGzGXgHphSbwb4X1UdeK+8slZBSTd5BFLFgGT+Lafl+a6PBrfPeK6nX6udyopipPlPjGtYwMY0NaBgAeC/V9e4AtVqy+0Gm7DU3e4zRxxQsc5oc8M7xwaSGAnll2MBI8ikNonkrpTUyMG2pqpJ5Ih1/fOB98K99KJG0sTZjmQMAefXHNa5PhETtD3aWcyKjo5d4EnduDRjn1/7rVdnJr2Xl+yV0kJhf1HQ8vyXyuX/AMxjx8x/h16V/wCEmf0lO1VPDS00tTUSCOGJhfI89GtAyT9lRvT002tOJdvqrw2rrO/rh7Ud+8lhkOWAEdAD+S+pxuPK6Wj7QLFpqitQOe4Yd3+Jzi4/mStsqTO5WEUAiAiAiCv3a603LUNtWqIWF4pWGnl5cgN24fq5ZnZS1UJLGdIzuzJTmSopufSNzt20egJP3Wut0QnVFkkVb+1FUxT6yoqRocJYKRm5w6YLyRn81entEpR4ARRR8OKV8bpC6WV75N7cAHkMN9MAfXK79Rb3KRFUEQEQEQEQRt2l2x/9HLzK6IOcwwbXhuXMJmYMj7qL+yC5kerLtAHHe+37yAeRDZGgZ9eZV4/KjysyiokRAVNu1dq2S/64bZaZ0jqG2zmLm7LDN7rXbfLxCmESkXspaBnt9IdTXCla2OdpNIc8855ux9/srAzRRTxOimjbIxwwWuGQVFo2mJlFPEThDSXJstfYJBTVW/e6BwJa8eIB8CoVey+aZucgzUQVdPMMAZZJH45/JfHdW4f7nkjLT1P9pdriZvr0mlvaY9GcYIXWmSK/xSurYG8nsjwJPn5FcXxF4sV95E9stVXJDD3f7RkLDucD1GR4YV8nWr5sX06x5+ZMfTorebW9fDA0Bw3vWpq0y1TKikpzGA+aZhAcB+6PPqrBaO0dYtK0zY7VShkmza+Z3N7+ecle/pHA7I+tf38PLzeT3T2V9OhUfdoOwtv3DKti3FslJIypjw3PMZaf/a9y+ghz1RuFN6msWs6bvK2WlhbMJPNmRzBx8wr6UFSytoKesi/s54myt+TgCP1UyiJifT2RVSIgIgKN+0Fe4rVpCGlc9olrahrGg9NviT6cwrU/NCJQ9wg0sb7M2vmieH+1QR7TLtd7O8kEgY8AD+StSr5Z86RWNCLJYRARARAUBdrm0P8AwulvTIpHxkCmft6byfdyr0nVkT6cD2U9SMoNdQUE1XHDT10Lo2h5xnqWNH5fkrdpk9kCKiXIcXtHO1xo59mhqW09Q2eOoge/4d7T0dyPLBPgqn2K437hrr2Vs/8AVp6eeNssTiQ2YAnr03N6c/mtaT40iVx9Jals+qLRDcrRWRzxyN5tDveYcAlpHpkLcLKY0kXy6ONzw9zGlzehI5hB41FDSVD4pJqeN7oX74yRza7IOR9QFkKIiIBFIOIaCSQAOZJ8FVztU64pZdaUun2U1LcaW3Ukhkim96P2mVuGv2/C/Y3z6ElXpHlFpmPTiuCllmv2o7ZRilfLFT1MRmcMgPYx4JDyOgPRXYVsnxBCvvaYqPbLzFQkER08TRvA6F3vEfbC6Hs40bIqB8rZXPLYcHPq7K+Y339Q/r/iHXn8PF194Z/aT1VBZeHlxs8UsjLhc6V0cRYduxhIa4k+u7bjxyuA7J+lnMqBeZotwp2Ow9/PJdyGP/cvp6xqsy5E+1kUWSREBEBEBEHP8RbGzUWirrazF3kslM804BxiUNOw/fCp/wAKtQV+kNd22oEjA6OQw1DHZIfGRg9AfQ8lrTzGkSu1Q1UNbRw1dM8PhmYHscPEEZXsskiqt2hmPl4pV4jcQQ2FpP8AD+xZ/v8Amr09on0sXw8o4aHRFngp3F0ZpWSBxGCd43fzW+VZSIoBEBEBEBEEddpKB1RwXv0bf/67iMZyBURH+ShTslTzQcR5oWsD2T0MsbyX/CAWHcB82gfVXj8qPlbBFRIiDT60vUWntM193mcGtp4JHgnoCGEj8wqU8M7J/S7iFQ0t1ZK6Cvqw+baQDzOX4znyV661KF5LTb6S1W2C3UELYKWnZsjjb0aFlKiWovGpLNag/wBrrYw9j2sdG0gvBPQYUB8WtbRXx87KaGJndP8AckEY34Hm7xC+Y671DHNZ41fM/P8A393W6fx7Rb6lvEONkoL7MKqSWmqGBgBmlLOXzytxpSu/o/X1E1Vbo3mrpe6cQeeMgnH2XzWDL+7ZIvr06uTHGSsxCxWl9YWC9x00FFVsZUSM92md8YwASPpldGv0XjcinIxxek+HzGXFbFbttAvmSNksZjkY17HDBa4ZB+i3ZqEcULP+A6/udBM3u2x12WAjZiN4BH+3L1Vv+CuoZb9pNoqX0zqin2gima0Rta5uWtG3ly5j6K0qxp3SKqwiAiAqqcZb9NrHVtJHE1vsrJn08OJWnvWbzjl4cwfVa4o87RM6TZwZsFJQ6ahqTTFsu79nvByxu0Lv1W87sVncbEVEiICICIC0PEGwDVGj7hY9zGuqWANc8ZAcHBw/RTE6kUyhhuGhuJQja90UtDMDE5zCOXUYz1BV3NOXBl1sNDco3h4qYGSEjpkjn+eVpkRDPRZJFGPHDhVR68oRXUMjaO/U7QIJtoLZWjPuPB8OfXqMfRTWdTslV/S+otY8M9V1cdNIaOc4FTE9jD3jAeQ9/wDkrV8NeLmldZsgo46sUd4dCx0tHMNmXbQX9248ngE/P0Wl678o2kNFkkRARBGHGbizZtF2mtoqaohqL+R3cFLgkNJbne84xgfw9SqeBtfqO+yvk/aVFRPvOwc95+Xh1+y2pGoVmPK4XZ80JJpLT0tRXQmKtqyMs8AwAY+uVKCzvO5TCtXGOubX6nvMO57XNn7oAsx8AA5H6KYeDNrjt+g7fKWvE1RC2SRz8ZP2+/1XzXAp3829vtv/AC6PItrDWPvpWnipqGXXnFCofSvkqKSnm7iliHvs7uN2N2BnqQ859QrVcP8AT9Jp3TNJRU0Za90TXzEnJLyBlfT28ViHO8OgRZpEQEQEQEQFUntHaepdJ8QqKvtcs0clwbLUyHAwx7pDjbj4Rzx9Fek+UT6WG4N3tt+4e22rMkbp4mez1DWDHdyM5Fp9eh+RC7BVn2mBU94hz1F04i6i3kCU3KaCIhufdY/YPnyCtjjcolbyhZ3dFBHv37Y2jdjG7A6r2VEiICICICICIOB7QwkPBrUZjdtc2CN5PkBKwn8gVAHZfbKeLVI4RHnFO5+0ghje7dzOPUgfPCvX8qJ9reoqJEQQD2udUOp6a1aQg35rM1VSW/wDIY36kP8AssLstaOgla/VFxke2emncyljOAHN2/H5+KvuIqjW5TZdtUUVHuZERJI0kHPIDCjHU/ESpkkeBU7GxvA7tnTP818R1nrN8szgwTqPmfn+jvcDgRrvye3O2ix6q1lUzTW+GaCmnmBkqZmd2zHn1yfHopC0zwbsdA5s15qprvKOZY9oZHn5DmfqVbo/RO+Pq5o8fEff9Tnc+K/gxpGfRUj45I3U0RbI3a8bR7w8io+1zoCnkpZqi007TmJzTB5ePurtdV6ZTlYNVjVq+v8Ao53E5VsOTc+pQp7FdbVcmT0dY8VDASCz3HwcumRz6KSeGvFl/wCztupZGPjYA0VgGHN/xDxHqF870vmTwsvbP5Z9/p+rr8zjRnx90e4TLSVMFXA2emmZLE7o5hyCvVfcxMTG4fOTGle+13pNj7ZT6tjkYNk0cM0ZjJzyd7+R0wB4rX9kXUIbX3K0VNRGyKSNklOHdXOLsYH3V9eFflZRFVYRARBw/GjV0ek9ISvY/wDr1YHQ0zB1Jx7zh8sj7hQXwL03+NXukqZnvexjxj3iQ9g+M81vj8V2ytNu+NelqKWnhpYWw08bY429GheiwaiICICICICIK4drqzVrb5adRRNaaX2X2WQlp91we4jn67/yXU9mDVjLhp3+jVRU95U0kZmhBbg92XkOH0JH3Wvuivnf6JmRZLCION4n8O7Fru2OZXwNjuMcRbS1jeT4jzIB8255kKn+tdG6h0LejRVsM0Ugy+GaLo5vmx3/ADp4LXHPwiUgaS7SGprVBJT3y0012iiihjhkZIYnjaMPLzg7ieR8PFSvZe0Lw6q6andcbhPbZ5KdksrTTSysjeese5rOZHnjCi1PsbdnauIeiLpTU9RQ6ntsjKg4jaZdj8+rXYc36gLJbrXSbpjD/SG3iQZy10wb0GT19FXskmdOavvGPRlA2OOkrfxCqlJEccfusGPF8h5MH5+ihfXHHvUVzEEVppYaGJlUx5HvOErQc7X9Dhw+SvGPXtEzuET220XfVF4mbFTSV1ZVSF22KEyE9TyHXAVo+CvBqm0jUR3u89zPdAwsZEGhzIfXPiev3VskzX37I8phWJea+K2Wuor5i0NhYXczjJ6AfU4H1XmvaKVm0/C8RudK0sppNVa2bSuDnPqKrD37j7mTzx9FM/FTUNNpDQ81DSUVVUTSUMkNPHTtz3TQzb3jyPha3I5rjdEr3Ra8/MvZzZjcRCE+yxpqO66iqbxV0LqqjZTOaHyB7GseSNpHgTgEYz45VpgAAAAAByAC72SfLwwIqJEQEQEQEQFG3HzRMWqtLvro393V2yCaZmGAmUBhcGfVzR9yprOp2Iu7KupXw38WLviYrix8xYGkhsjGD+Q6/JWaVrxqUQKmVIDUazNZVP3j8TzMxnMyDvQTz8z0U450TEz6XNHIYCLNIiAiAiAiAiDi+OjSeD2q3hod3VsmmIPkxu8//FV97LT3x8VIA2F7XS0k3eZHwt255/XC0r+WULbIs0i8q2ojpKOeql/s4Y3SO+QGT+iCll1ra3iBr81UEM7p62c+zxn91mTgf3euT8ypdor3Q6YsAtlFKZmxP2GRgxk//a+b/aDk2ikYae59un0zjxNrXt6c1V3e5XuWGjtkNQ+WWQ45Ek59F32guETmH23Vksj5my7200crXMd6vOOfyXI6T0qc+Tuv+WHv5vJrip21nzKX4IooIWwwxtjjYMNa0YAC+193EafOiIOc1NpC13p4qTH3NUxu1kjR4eSibVXDmqtkbp2xiRoz+1aw4HzC+Y6z02d/Xxf1dngc31ju0mjNYXrRtQ2GRz5aLvD3kTskHnzx6qfdMaos+oqYS26qa5+Muidye35hX6Hz4tX6F58/H/T+inUeN2z9SsfzePEXT0eqdGXKxyNafaYsM3DIDhzH6KoWmqqbS3FOq2QikAqn0xBjwIBnw5npjzX01XI7drr0E4qqCnqQQRNE2QY6cwCvZVSIgIgqpxl1FV6015T0DKeeA0ULoIaPvc5e85c8jzwzy6KbeCOmG2DRlH7RQtgrHguzkk7XAHx6f7ALa0dtdKRO5d8ixXEQEQEQEQEQanWVojv+lLpZpRltZSvix6kcvzwqp8M/xTSnF2joZWT0NRESx7ZYXEYd7uC1uXYP8srXH5iYRK4SLJIiAsW426guUD4K+jgqonsdG5srA4Fruo5/JBE2s+z1o+8F01mfLYpu7DQyEb4iR0JB5/YhRtVdmnVsDp3UF009NvjDGmd0ge3Dt3IiPryaPutYyfdTs87cpqLRuuNHXKhdUWGpmbHP3bJ4oPaIS/HLqOeVzVBp/VNTUwR/gVfNUd4SwGiecgdcjZ5/Rax5jx6S30XDjiDPVwTizySMqZHRgxs6P59QOQH28FJWguAlzlmFRqMx0rRKHs97vCWY+BzOQ6qtrRCU+6W01Y9MUJorHboaOJx3P283PPmXHmVt1hM7SLgOOFx9l0zFRB+w1MoJP+H3gP8AVtXj5+T6fHvP6NMUbvDleANhiqK+t1DM973CZ3dAnk3K5rtTVzq/WNv09CH74rf3+Wjdue+QgM9DhuceO4LPo+PWGJ17X5dt3lMHCfSVNo/SNNb4XSPlexr5nSM2uyee3HhgkhdcuhM7nbCBFAIgIgIgIgL4niZPBJDIMskaWuHmCMFBUaGlrOE3FP2Wnqt4pCBvaM95CcPEZHhubgfNW2oKhtXQwVTMbZo2yDHkRlaZPOpRD5ukogtlVO5z2tjhe8lgy4YaTkeqrNwl0vNdNQ0jqqJ8sccjDPHjOACMk/VKTqJlK0KLMEQEQEQEQEQc1xWbG/hfqtsrdzDZazcPTuXqtHZxlktvGSgp6in3PnppI2tYdxj3M3bvy5+XNXr6lC3iKiRRT2lNZs05o2SxxMlNde4JIYXs6MaCwPJ/yvKre3bWZTEblXfh4+ptdbHWxxuM0QPd45FhIxz+6k7Q/D+6X5ntNwMlPRd6MvORknyGOfNfMZsF+VyPw/LsY88YqbTtpzS9msEIjoKbmDnvJDufn5rdL6TDhphp2Ujw5OTJbJbusItVBEBfMjGSMLHtDmuGCD4pMbInTkNU6BtV2ixDGyI7923by+nkolv+lb7p64VD6MTsjiLXNexxBd/e5fLC+U6j0qcNvq4fTscblxkr9O7aWXjDcLbDNT3mH2l0bMNc05fu/mou4601urrs3VtjuUNTSVD2iSPYWyRyhnj9sZXR6f1G86jK8ufjxSfwSnrs76qdf9Gx0tVK51VSjaN73PJYOQ5ny5BScuzFot5h4daEUgiDylpqaWZk8lPE+WP4HuYC5vyPh1XqgIgIgIgIgIgIgLwmo6SaojqJqWCSaL+zkfGC5nyJ5hB7ogIgIgIgIgIgIgKvnFW+m/ayZRUL5JxTVLYw0Zw4NPPA/iyeXmub1Sd4or95ejjx5mUy6EshsViZSyMayVzi54Ds48gsypsNnqbs27T2+GSua0MEzgdwA8F7sFPpY4r9mFp3O2yRaIEQEQEQEQEQEQQ7xv4XXnV+rLdfbJU07CyjNNUsnOANry5hbyzk948H5NUn6TpbhQ6ZttHdZY5q2CmZHM9nQuAxyWlrxNYgbGRjJGFkjGvaeRa4ZBWm01pm22Pvn0sDRLLK5+7qQCeQ+2FSJ1Ghu0UAiAiAiAiAiDxraaKso56Sdu6GeN0cg82uGCPsVBGi+Et5sHFG3XWCkDLZQPJY6SZpcQWlvh/9q1ZiInaE+oqpFEPFrhXctb6zZdW3PuKeG391AJBvYx+SSA3cMbuWXfLyVb17o0tWdTtvtE8LbPp4Uk1RIa6qhYMlzcM3YwcD7dfJd+1rW52gDJyceapixVxx4LWm0v1FqqIgIgIgLxrKWCrgMNTE2SM9WlRasWjUpideXC644b229U9SaWlgjmlcxwIGC3BG4Dw5gKFOIXDLVNuju8woIZaGGNjxM1/ukbsBrW9S7muV/psVvuvp6/3nurESx+y7c6u08Q47a9j46apifDMx+QWOPvMdj1c0j6q2a6eOnZXTyTMTPgRXQIgIgIgIgIgIgIgIgIgIgIgIg5/UestPWGImuuEZl6CGM7pD9PD5lY2j9a0WpKx9PBTSwYi7xrnkEO97BHzHL7rwz1DF9WMUTuZeiONk7Jvrw6lF7nnEQEQcTxe1V/Rmx0/dTllTVSFrWt+MsA5lvyc5g+qjzgxpmur7tFfHExx08gdvcM7sEe788Lkcrvy8utK+o1L045iuO0z8p5Rdd5hEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEHPHRunm6gF8gtdJDXdwYDKyFocRnIJ83DzOV0KbRERAiJEQEQEQEQEQEQEQEQEQEQEQFy/EmvfQafndFUPhllgkihLeneHG0n5c1hyb9mG1v0Xxxu8QgE0jrlcRTse/4A3lzJPTkpm4WaKq9Pf1mukjLu7xG1rsnJ+In3R6fmvlekca+bPF59V8uxy8tKYu2PcpARfYuIIgLU6n1HaNOUJq7pVxxf+nHn35D4BrepKre0UibSIWstHXcVOIktwuIlZa6QMD9nKNsY5iIf3nk5J8vkFPNDSU1DSx0tJCyCCMYYxgwAF5+NHdvJ91rbiO2Xsi9SoiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAiAuA43Rb9PwPeXuYyQnY3xO0jP5rxdR/8NdrgjeSIRbpiVlvv9O+bGJG5PmDlWCsNbHXW2KZkgece96FcTofIpGWcfzMPbzcU1rFmei+oc0RByetNd2bTlMc1EdRVHk2Jh3YPrj9Oqge41N51nqN9VWVLpHuxHE1rcbG5yGY+p9ea+e6lz+630qeo9ujx+Jbt77LDaEsFLp+wRU9PHtkmAlncRgl5A6/Lot8u7ir2UirwXnutMiLRURARARARARARARARARARARARARARARARARARARARARARARARARARARARARARARARARARARARARARAWvv9qprvb5KWoYDuY4MJ/dJGMqmSkZKzWflalppaLQgDUllu9juT/xKmkjBH7OVvvh/P8AXksLT2oqmiru+L5BKOTCOY/59l8ByONk42WZ15j0+twZ8WfFqZ8u601xMrYKiH8RPtVHKTna0b2DHLBz73PA+q66PiZphxa10lTG5zcjfCQPlnp5/ZdrhdeiuPtzxMzH2cnl9Nnv7sXp913EXTkFI2VlYx8j4y4NHvYPkcKONR8U6qr200NYyAvJGYsjHl4rTldVnNHbh9f3/oyw8Ltn/cY+keHOoL62Ctq6vuKZsme8qN2+QeYb+8PmQpX0joi16emlnYTUyybcGRo9zHl6+q34HTO2Yy5Pf2U5PLnzjr6dSi7rnCICICICICICICICICICICICICICICICICICICICICICICICICICICICICICICICICICICICICICICICIPOeCGeN8c0bXse3a4EdR5LgdQcKLDXzuqaCWW3SlmNrAHRud5lp/lheXlcWnIrqy+O80tuHB1/C/VVBPCykpo64ZJM0U4bsx05OwtVDw61zOCyO2VEeGkZmmja3ryxzz+nVfP36Hl7vGtfo69OoUivltLfwf1NXxym5TU9FIxoEOZd3P/LnkpG0RwysOnHxVkzXXC4saP20vwMd5sb4fM5Pqujw+l/SneSYnTzcjm924p8u5RdlzhEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEBEH//2QpBZ2FtYSBsb2dvIGRlc2lnbmVkIGJ5IFRhdGlhbmEgTW9yb3pvdmEK"
    }
   },
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "![logo](attachment:agama.jpg)\n",
    "# Tutorial for the Agama stellar-dynamical framework "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import agama, numpy, warnings, matplotlib, matplotlib.pyplot as plt, time\n",
    "%matplotlib inline\n",
    "numpy.seterr(all='ignore')\n",
    "warnings.filterwarnings('ignore')  # disable overflow warnings\n",
    "plt.rc('figure', figsize = [8.0, 6.0])\n",
    "plt.rc('figure', dpi = 75)\n",
    "plt.rc('font', size=12)\n",
    "plt.rc('legend', fontsize=12)\n",
    "plt.rc('mathtext', fontset='stix')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Unit system\n",
    "By default, `Agama` works in the so-called \"N-body units\", which are defined by the convention that G=1. This is convenient for many theoretical applications, since stars in a stellar system with a typical mass ~1 and typical size ~1 will have typical velocity ~1. Of course, gravity is a scale-free interaction, thus such a model can be scaled to physical units to represent anything from a globular cluster to a giant elliptical galaxy without changing anything in the internal dynamics. For example, if one sets the length unit to be 1 kpc and the velocity unit - to 1 km/s, then the time unit is 0.98 Gyr and the mass unit computed from the condition G=1 is $2.325\\times10^5\\,M_\\odot$.\n",
    "\n",
    "However, often it is more convenient to set all three base units (length, velocity and mass, or length, time and mass) independently, which means that $G\\ne 1$, or even employ different (but compatible) units for the same physical quantity in different parts of the code (e.g., km/s/kpc and mas/yr, both of which are units of angular velocity). \n",
    "`Astropy` provides a very general unit handling subsystem that can address these needs, but it comes with non-negligible overhead costs and requires to write quite verbose code. By contrast, `Agama` implements a much simpler approach, in which the user can select any triplet of base units, and all dimensional quantities passed as input or produced by library routines will be automatically unit-converted inside these routines. This is achieved by calling\n",
    "```\n",
    "agama.setUnits(...)\n",
    "```\n",
    "at the beginning of the script (and only once). The arguments of this routine are expressed in terms of fiducial \"standard\" units - length in kpc, velocity in km/s, time in Myr, mass in Msun (exactly two out of the first three need to be provided). For instance, the \"Gadget\" units commonly used in cosmological simulations are defined by\n",
    "```\n",
    "agama.setUnits(length=1, velocity=1, mass=1e10)\n",
    "```\n",
    "and if for some reason one wishes to adopt the distance from the Sun to the Galactic center as the length unit and the local circular speed as the velocity unit, then\n",
    "```\n",
    "agama.setUnits(length=8, velocity=220, mass=1)\n",
    "```\n",
    "does the job.\n",
    "\n",
    "This routine also accepts `Astropy` units and quantities as input arguments, but this is the *only* exception: all other methods and routines work with plain `numpy` arrays, and the meaning of these physical quantities is determined by the initial unit setup, and conversions between compatible units remain the user's responsibility. For example, if length unit is 1 kpc and velocity unit is 1 km/s, the unit of proper motion turns out to be 0.211 mas/yr.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "units: {'length': 1.0, 'velocity': 1.0, 'time': 977.792221683525, 'mass': 1.0}, G=4.30092e-06\n"
     ]
    }
   ],
   "source": [
    "agama.setUnits(length=1, velocity=1, mass=1)\n",
    "print(\"units: %s\" % agama.getUnits() + \", G=%g\" % agama.G)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Potentials\n",
    "`Agama`has a number of built-in potential and density models (listed in Table 1 of reference.pdf for the list of models and their parameters), and could be augmented with arbitrary user-defined models, as will be discussed later in this tutorial.\n",
    "\n",
    "One may create a simple potential by providing a list of parameters, or a composite potential from existing objects or from several groups of parameters, or load parameters from an INI file (which may contain several sections with parameters for each component).\n",
    "\n",
    "Of course, individual components of a composite potential can be indexed, and may themselves turn out to be composite potentials of arbitrary complexity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CompositePotential{ Dehnen, MiyamotoNagai, NFW } (symmetry: Axisymmetric)\n",
      "bulge: Dehnen (symmetry: Spherical), pot[0]: Dehnen (symmetry: Spherical), they are identical: True\n",
      "CompositePotential{ Dehnen, MiyamotoNagai, NFW } (symmetry: Axisymmetric)\n"
     ]
    }
   ],
   "source": [
    "pot_names = ['bulge', 'disk', 'halo']\n",
    "# 1st approach: create separate potential components, then combine\n",
    "pot_bulge = agama.Potential(type='Dehnen', scaleRadius=1.0, gamma=1.0, mass=2e10)\n",
    "pot_disk  = agama.Potential(type='MiyamotoNagai', scaleRadius=3.0, scaleHeight=0.3, mass=5e10)\n",
    "pot_halo  = agama.Potential(type='NFW', scaleRadius=15.0, mass=5.5e11)\n",
    "pot_total = agama.Potential(pot_bulge, pot_disk, pot_halo)\n",
    "print(pot_total)\n",
    "# extract a single component from a composite potential\n",
    "print(\"bulge: %s, pot[0]: %s, they are identical: %s\" % (pot_total[0], pot_bulge, pot_total[0]==pot_bulge))\n",
    "\n",
    "# 2nd approach: create a list of dictionaries with parameters and feed it to the Potential constructor\n",
    "pot_params = [\n",
    "    dict(type='Dehnen', scaleRadius=0.5, gamma=1.0, mass=1e10),\n",
    "    dict(type='MiyamotoNagai', scaleRadius=3.0, scaleHeight=0.3, mass=6e10),\n",
    "    dict(type='NFW', scaleRadius=16.0, mass=5.5e11)\n",
    "]\n",
    "pot_total = agama.Potential(*pot_params)\n",
    "print(pot_total)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `Potential` class provides a handful of basic methods, the most fundamental being the computation of density, force ($-\\partial\\Phi/\\partial\\boldsymbol{x}$) and force derivatives. Nearly all methods of this and other classes can accept single points or arrays of points as input, and in the latter case, the computations are internally `OpenMP`-parallelized if the array size is large enough to justify this.\n",
    "\n",
    "Positions and velocities are always given in Cartesian coordinates, so to evaluate the density or potential along the $x$ axis for an array of points $r$ of length $N$, one needs to provide a 2d array of shape $N\\times3$: `xyz = numpy.column_stack((r, r*0, r*0))` or something equivalent.\n",
    "\n",
    "Most high-order \"convenience\" operations can be expressed in terms of the basic ones, for instance, the circular-velocity curve defined as $v_\\mathrm{circ}(R) = \\sqrt{R\\, \\partial\\Phi(R,0,0)/\\partial R}$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Vcirc at 8 kpc = 227.099 km/s, Vescape = 556.364 km/s\n"
     ]
    }
   ],
   "source": [
    "def scalarize(x):\n",
    "    return x if len(x)>1 else x[0]\n",
    "\n",
    "def v_escape(pot, r):\n",
    "    return scalarize((-pot.potential(numpy.column_stack((r, r*0, r*0))) * 2) ** 0.5)\n",
    "\n",
    "def v_circ(pot, r):\n",
    "    return scalarize((-r * pot.force(numpy.column_stack((r, r*0, r*0)))[:,0]) ** 0.5)\n",
    "\n",
    "gridR = agama.nonuniformGrid(50, 0.01, 30.0)\n",
    "ax=plt.subplots(1, 2, figsize=(12,5))[1]\n",
    "ax[0].plot(gridR, v_circ(pot_total, gridR), label='total', color='k')\n",
    "for i in range(len(pot_total)):\n",
    "    ax[0].plot(gridR, v_circ(pot_total[i], gridR), label=pot_names[i], color=['r', 'g', 'b'][i])\n",
    "\n",
    "ax[0].set_xlabel('R [kpc]')\n",
    "ax[0].set_ylabel(r'$v_\\mathsf{circ}$ [km/s]')\n",
    "ax[0].legend(loc='lower right')\n",
    "\n",
    "ax[1].plot(gridR, v_escape(pot_total, gridR))\n",
    "ax[1].set_xlabel('R [kpc]')\n",
    "ax[1].set_ylabel(r'$v_\\mathsf{escape}$ [km/s]')\n",
    "print('Vcirc at 8 kpc = %g km/s, Vescape = %g km/s' % (v_circ(pot_total, 8.0), v_escape(pot_total, 8.0)))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Of course, any potential also provides a method for computing the density, as in the following example.\n",
    "In fact, one may create density-only models without an associated potential (instances of `agama.Density`), although, of course, it is more interesting to feed these into the Poisson equation and obtain the potential – we'll get to this shortly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "solarRadius = 8.0  # kpc\n",
    "ax      = plt.subplots(1, 2, figsize=(12, 5))[1]\n",
    "gridR00 = numpy.column_stack((gridR, gridR*0, gridR*0))    # for the radial profile at z=0\n",
    "gridz   = agama.symmetricGrid(50, 0.01, 10)\n",
    "gridR0z = numpy.column_stack((gridz*0 + solarRadius, gridz*0, gridz))  # for the vertical profile at solar radius\n",
    "for i in range(len(pot_total)):\n",
    "    ax[0].plot(gridR, pot_total[i].density(gridR00), label=pot_names[i], color=['r', 'g', 'b'][i])\n",
    "    ax[1].plot(gridz, pot_total[i].density(gridR0z), label=pot_names[i], color=['r', 'g', 'b'][i])\n",
    "ax[0].set_xlabel('R [kpc]')\n",
    "ax[0].set_ylabel(r'$\\rho(R)$ at $z=0$ $[M_\\odot/\\mathsf{kpc}^3]$')\n",
    "ax[0].set_yscale('log')\n",
    "ax[0].set_ylim(1e5,1e10)\n",
    "ax[0].legend(loc='upper right')\n",
    "ax[1].set_xlabel('z [kpc]')\n",
    "ax[1].set_ylabel(r'$\\rho(z)$ at %g kpc $[M_\\odot/\\mathsf{kpc}^3]$' % solarRadius)\n",
    "ax[1].set_yscale('log')\n",
    "ax[1].set_ylim(1e3,1e8)\n",
    "ax[1].legend(loc='upper right')\n",
    "None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "By integrating the density profile along the line of sight, one gets the projected density. The orientation of this line of sight w.r.t. the model coordinate system is specified by three Euler angles $\\alpha, \\beta, \\gamma$ (unfortunately these letters are also used in different contexts, e.g., to denote the power-law indices of the `Spheroid` density profile). See Section A.3 in reference.pdf for details.\n",
    "\n",
    "The most important in practice is the inclination angle $\\beta$, which ranges from 0 for a face-on orientation (with $z$ being the line of sight) to $\\pi/2$ for an edge-on orientation (line of sight is in the $x-y$ plane). In the latter case the other angle $\\alpha$ specifies the direction of the line of sight within this plane; $\\alpha=0$ puts it along the $y$ axis and $\\alpha=\\pi/2$ – along the $x$ axis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "grid2R, grid2z = numpy.meshgrid(gridR, gridz)  # two 2d arrays of shape (len(gridR), len(gridz))\n",
    "gridRz = numpy.column_stack((grid2R.reshape(-1), grid2z.reshape(-1)))  # array of shape (len(gridR)*len(gridz), 2)\n",
    "gridR0 = numpy.column_stack((gridR, gridR*0))  # array of shape (len(gridR), 2)  for the 1d profile\n",
    "plt.figure(figsize=(12,5))\n",
    "ax=[plt.axes([0.08,0.1,0.32,0.8]), plt.axes([0.48,0.1,0.5,0.8])]\n",
    "for i in range(len(pot_total)):\n",
    "    # face-on, 1d profile\n",
    "    ax[0].plot(gridR, pot_total[i].projectedDensity(gridR0, beta=0), label=pot_names[i], c=['r','g','b'][i])\n",
    "    # edge-on, 2d contour plot\n",
    "    Sigma = pot_total[i].projectedDensity(gridRz, beta=numpy.pi/2).reshape(grid2R.shape)\n",
    "    ax[1].clabel(ax[1].contour(grid2R, grid2z, numpy.log10(Sigma), levels=numpy.linspace(0, 10, 11),\n",
    "        cmap=['Reds', 'Greens', 'Blues'][i], vmin=3, vmax=10), fmt='$10^%.0f$')\n",
    "ax[0].set_xlabel('R [kpc]')\n",
    "ax[0].set_ylabel(r'$\\Sigma(R)$ [$M_\\odot/\\mathsf{kpc}^2$]')\n",
    "ax[0].set_title('face-on surface density')\n",
    "ax[0].set_yscale('log')\n",
    "ax[0].set_ylim(1e5,1e10)\n",
    "ax[0].legend(loc='upper right')\n",
    "ax[1].set_xlim(-0.5, 30)\n",
    "ax[1].set_xlabel('R [kpc]')\n",
    "ax[1].set_ylabel('z [kpc]')\n",
    "ax[1].set_title(r'edge-on surface density [$M_\\odot/\\mathsf{kpc}^2$]')\n",
    "None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Optimized potential construction\n",
    "In the previous example, we used three distinct Galactic components and they correspond to three elements of a composite `Potential` object. Let's now try another Galaxy model, with a Sersic bulge, an exponential disk, and a truncated NFW halo (the 'vanilla' NFW is a rather awkward model with a logarithmically diverging total mass). The latter one is represented by a Spheroid model with $\\gamma=1$, $\\beta=3$ and $\\alpha=1$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CompositePotential{ DiskAnsatz, Multipole } (symmetry: Axisymmetric)\n"
     ]
    }
   ],
   "source": [
    "pot_params = [\n",
    "    dict(type='Sersic', scaleRadius=1.5, mass=2e10, sersicIndex=2.0),\n",
    "    dict(type='Disk', scaleRadius=2.8, scaleHeight=0.3, mass=5e10),\n",
    "    dict(type='Spheroid', scaleRadius=20.0, mass=1e12, outerCutoffRadius=250.0, gamma=1, beta=3, alpha=1)\n",
    "]\n",
    "pot_total2 = agama.Potential(*pot_params)\n",
    "print(pot_total2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We see, unexpectedly, that the resulting potential contains two rather than three components, and their names bear little resemblance to the original ones.\n",
    "\n",
    "This is because `agama.Potential` constructor rearranges the input density models and creates a more computationally efficient potential object. In fact, the chosen components are pure density models without an explicit potential counterpart. (You may be surprised why NFW used to have one in the previous example: when we specified an exponential cutoff, it no longer has an analytic expression for the potential, and we had to replace it by a more general Spheroid density-only model. Note that also the `mass` parameter of the NFW model is not the total mass (which is infinite), but the mass within $\\sim 5.3\\times$ scale radius – see footnote in Section 2.2.4 of reference.pdf; but for the Spheroid model it is really the total mass).\n",
    "\n",
    "To explain what happens under the hood, let's first create another composite potential with all three galactic components represented separately. This is achieved by iterating over the list of `pot_params` and creating a separate `agama.Potential` for each dictionary from this list, then combining the resulting array into another, top-level instance of Potential:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CompositePotential{ Multipole, CompositePotential{ DiskAnsatz, Multipole }, Multipole } (symmetry: Axisymmetric)\n"
     ]
    }
   ],
   "source": [
    "pot_total3 = agama.Potential(*map(agama.Potential, pot_params))\n",
    "print(pot_total3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Again the names of these components don't coincide with the input parameters, but at least there are three of them.\n",
    "\n",
    "The thing is that analytic expressions for the potential are available for some of the simpler density models, but in general this is not possible. Instead, `Agama` uses two very powerful general-purpose methods for solving the Poisson equation, i.e. computing the potential for the given density profile. The one used here is the `Multipole` expansion, in which the density is first decomposed into spherical harmonics with radially-varying amplitudes, then each harmonic term produces its own term in the potential, and the result is synthesized back (much like solving the Poisson equation in Cartesian coordinates via the Fourier transform). \n",
    "\n",
    "Two of our input density components (`Sersic` and `Spheroid`) are spherically-symmetric, so they would produce only one multipole term ($l=0$). The `Disk` potential is computed using the technique introduced by Kuijken & Dubinski (1995) and further developed by Dehnen & Binney (1998): namely, it is split into a `DiskAnsatz` part, which is strongly concentrated near the equatorial plane and has analytic expression for the potential, and the residual density profile which is not strongly concentrated and is efficiently represented by spherical harmonics.\n",
    "\n",
    "You now see why in the previous attempt we obtained only one `Multipole` and one `DiskAnsatz` component: it makes sense to combine all three density profiles into just one instance of a Multipole potential, and that's exactly what happens under the hood. We may illustrate the gain in computational efficiency by comparing the time needed to evaluate the potential for a large number of points in the previous (automatically optimized) case with the manually created three-component potential:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "optimized potential: 3.87e+07 eval/s, separate components: 2.38e+07 eval/s\n"
     ]
    }
   ],
   "source": [
    "npoints = int(1e7)\n",
    "points = numpy.random.random(size=(npoints,3))\n",
    "t1 = time.time()\n",
    "pot_total2.force(points)\n",
    "t2 = time.time()\n",
    "pot_total3.force(points)\n",
    "t3 = time.time()\n",
    "print(\"optimized potential: %.3g eval/s, separate components: %.3g eval/s\" % (npoints/(t2-t1), npoints/(t3-t2)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "While the difference is only about a factor $1.5\\times$ in this case, it may be more significant if the bulge or the halo were non-spherical, necessitating more expensive higher-order Multipole expansions if instantiated separately.\n",
    "\n",
    "On the other hand, with the optimized potential, we cannot plot the circular-velocity curve for galactic components separately, so the manually created variant is also useful sometimes.\n",
    "\n",
    "One should also keep in mind that each `Disk` component creates its own `DiskAnsatz`, which has non-negligible evaluation cost; the popular McMillan(2017) model has four of these! Is it possible to do better – namely, have just one potential component for everything? Yes, but it may not be a `Multipole`: the strongly flattened disky components are poorly represented even with a high-order Multipole expansion.\n",
    "\n",
    "There is another Poisson solver more suitable for this case, named `CylSpline`. It works by Fourier decomposing the input density profiles (in the axisymmetric case, there is only one $m=0$ term in this expansion), solving the Poisson equation on a rectangular 2d grid in $R,z$ for each Fourier term, and interpolating the result for any point in space between the grid nodes. It is more costly to construct, but comparable to `Multipole` in terms of evaluation efficiency, and greatly exceeds it in accuracy for flattened systems. The only limitation is that it is not working well with very cuspy density profiles, but the NFW and low-$n$ Sersic models should be relatively safe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CylSpline potential took 0.130 s to construct, runs at 4.19e+07 eval/s\n"
     ]
    }
   ],
   "source": [
    "t4 = time.time()\n",
    "dens_total = agama.Density(*pot_params)\n",
    "pot_total4 = agama.Potential(type='CylSpline', density=dens_total)\n",
    "t5 = time.time()\n",
    "pot_total4.force(points)\n",
    "t6 = time.time()\n",
    "print('CylSpline potential took %.3f s to construct, runs at %.3g eval/s' % (t5-t4, npoints/(t6-t5)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So yes, it runs a bit faster in this case, and the savings will be more significant for larger number of density components crammed into a single potential (as opposed to being treated separately).\n",
    "\n",
    "Let's check the accuracy of the potential approximation: in absence of the \"ground truth\" potential model, we are forced to look at the density accuracy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "ax = plt.subplots(1,2,figsize=(12,5))[1]\n",
    "for pot, label, color in [\n",
    "    [dens_total, 'input density', 'c'],\n",
    "    [pot_total2, 'Multipole+DiskAnsatz', 'm'],\n",
    "    [pot_total4, 'CylSpline', 'y'] ]:\n",
    "    ax[0].loglog(gridR, pot.density(gridR00), label=label, color=color)\n",
    "    ax[1].loglog(gridz, pot.density(gridR0z), label=label, color=color)\n",
    "\n",
    "ax[0].set_xlabel('R [kpc]')\n",
    "ax[0].set_ylabel(r'$\\rho(R)$ at $z=0$ $[M_\\odot/\\mathsf{kpc}^3]$')\n",
    "ax[0].legend(loc='lower left')\n",
    "ax[1].set_xlabel('z [kpc]')\n",
    "ax[1].set_ylabel(r'$\\rho(z)$ at %g kpc $[M_\\odot/\\mathsf{kpc}^3]$' % solarRadius)\n",
    "ax[1].legend(loc='lower left')\n",
    "ax[0].set_xlim(gridR[1], gridR[-1])\n",
    "ax[1].set_xlim(gridz[len(gridz)//2], gridz[-1])\n",
    "ax[0].set_ylim(3e5, 3e11)\n",
    "ax[1].set_ylim(3e6, 3e8)\n",
    "None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Wow, what are those wiggles? Didn't I promise that `CylSpline` is highly accurate for disky systems?\n",
    "\n",
    "Well, yes, but we may need to tune its parameters a little.\n",
    "\n",
    "By default, both `Multipole` and `CylSpline` use grids with 25 nodes in radius (and in $z$ for the latter) – this is usually sufficient to represent even models with a wide dynamical range, since the grid nodes are spaced logarithmically and the potentials are interpolated with 5-th degree splines. The grid extent (`rmin`/`rmax` for both, and additionally `zmin`/`zmax` for the latter) is also determined automatically using some heuristics, but sometimes it may be advantageous to adjust these parameters. \n",
    "\n",
    "To find out the automatically selected grid parameters, we export the potential coefficients into a text file and take a look."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Potential]\n",
      "type=CylSpline\n",
      "gridSizeR=25\n",
      "gridSizez=25\n",
      "mmax=0\n",
      "symmetry=Axisymmetric\n",
      "Coefficients\n",
      "#Phi\n",
      "0\t#m\n",
      "#R(row)\\z(col)\t0.0000000000000\t0.1465770018021\t0.3862145558085\t0.7779960534796\t1.4185164533555\t2.4656980714154\t4.1777268239967\t6.9767087762350\t11.552741717938\t19.034060940740\t31.265209192183\t51.261814482440\t83.954102640026\t137.40245997193\t224.78474681916\t367.64534040554\t601.20696869678\t983.05498458840\t1607.3352154941\t2627.9659127353\t4296.5867797455\t7024.6015311611\t11484.610770573\t18776.243781473\t30697.275580197\n",
      "\n"
     ]
    }
   ],
   "source": [
    "pot_total4.export('MWpotCylSpline.ini')\n",
    "with open('MWPotCylSpline.ini') as f:\n",
    "    lines = f.readlines()\n",
    "    print(''.join(lines[:10]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The long header line shows the nodes in $z$, which are by default the same as in $R$. \n",
    "Probably the innermost grid node is a bit too far, and the cuspy density profiles of both bulge and halo deteriorate the accuracy of interpolation in a few adjacent segments, while the outermost node is certainly too far given our halo cutoff radius at 250 kpc.\n",
    "\n",
    "We can increase the number of nodes and reduce `rmin` to better resolve the problematic inner region, whereas `rmax` may be safely set to 2-3 times the outer cutoff radius."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "pot_total5 = agama.Potential(type='CylSpline', density=dens_total, rmin=0.03, rmax=600, gridsizeR=30, gridsizez=30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "ax = plt.subplots(2,2,figsize=(12,10))[1]\n",
    "for pot, label, color in [\n",
    "    [dens_total, 'input density', 'c'],\n",
    "    [pot_total2, 'Multipole+DiskAnsatz', 'm'],\n",
    "    [pot_total5, 'CylSpline', 'y'] ]:\n",
    "    ax[0,0].loglog(gridR, pot.density(gridR00), label=label, color=color)\n",
    "    ax[0,1].loglog(gridz, pot.density(gridR0z), label=label, color=color)\n",
    "    if pot == dens_total: continue\n",
    "    ax[1,0].loglog(gridR, abs(pot.density(gridR00) / dens_total.density(gridR00) - 1), label=label, color=color)\n",
    "    ax[1,1].loglog(gridz, abs(pot.density(gridR0z) / dens_total.density(gridR0z) - 1), label=label, color=color)\n",
    "\n",
    "ax[0,0].set_xlabel('R [kpc]')\n",
    "ax[0,0].set_ylabel(r'$\\rho(R)$ at $z=0$ $[M_\\odot/\\mathsf{kpc}^3]$')\n",
    "ax[0,0].legend(loc='lower left')\n",
    "ax[0,1].set_xlabel('z [kpc]')\n",
    "ax[0,1].set_ylabel(r'$\\rho(z)$ at %g kpc $[M_\\odot/\\mathsf{kpc}^3]$' % solarRadius)\n",
    "ax[0,1].legend(loc='lower left')\n",
    "ax[1,0].set_ylabel('relative difference')\n",
    "ax[1,1].set_ylabel('relative difference')\n",
    "ax[0,0].set_xlim(gridR[1], gridR[-1])\n",
    "ax[1,0].set_xlim(gridR[1], gridR[-1])\n",
    "ax[0,1].set_xlim(gridz[len(gridz)//2], gridz[-1])\n",
    "ax[1,1].set_xlim(gridz[len(gridz)//2], gridz[-1])\n",
    "ax[0,0].set_ylim(3e5, 3e11)\n",
    "ax[0,1].set_ylim(3e6, 3e8)\n",
    "None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Well, that's much better now, with density accuracy at the level of a few percent for `CylSpline`.\n",
    "Remember that the density is a second derivative of the potential, so the accuracy of the force (first derivative) is much higher than that.\n",
    "\n",
    "One may be tempted to increase the grid size and resolution even more, but this probably won't help here for two reasons. First, the `CylSpline` potential at grid nodes is computed by numerical integration with a finite target accuracy ($10^{-6}$), so the density computed as the 2nd derivative of slightly noisy potential values has unavoidable numerical noise at a higher level. Second and also related, the interpolation becomes more noise-dominated for small grid spacing (and hence smaller difference between values in adjacent nodes, which themselves carry some error). It is inadvisable to reduce `zmin` below 0.1-0.2 times the disk scale height.\n",
    "\n",
    "In the above example, `Multipole` in combination with `DiskAnsatz` still fares better, and that's why it is the default approach suitable for most cases, but `CylSpline` is indispensable when one needs to represent a potential of a *custom* flattened model other than `Disk`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Custom density and potential models\n",
    "Whatever large variety of built-in models is, one may need something outside the box. \n",
    "\n",
    "`Agama` is designed to be extensible both at the level of the core C++ library (which few users seem to care about) and the Python interface. In particular, one can define custom density and potential functions, which can be used *anywhere* in the code in the same way as the built-in ones. Doing this efficiently, however, is another story.\n",
    "\n",
    "Let's start with an example of a custom boxy bar density profile – a triaxial Einasto model with equidensity contours being generalized ellipsoids: \n",
    "$\\rho(x,y,z) = \\rho_0 \\exp(-d^{1/n})$, where $d=[(x/a)^k + (y/b)^k + (z/c)^k]^{1/k}$. For $k=2$ this is a standard ellipsoid, and this model should be equivalent to the built-in `Spheroid` profile with $\\gamma=\\beta=0$ and `cutoffStrength`=1/k.\n",
    "\n",
    "We need to define a function that takes an array of shape $N\\times 3$ and returns the density for $N$ points in one call (even if $N=1$, the input array is still two-dimensional). This should be a standalone function (not a class method), and in order to make its parameters adjustable, it is convenient to define an enclosing \"factory\" function, which takes these parameters, creates and returns the actual density function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<function make_bar_density.<locals>.dens_bar at 0x12f951800>\n",
      "Sersic bulge mass: 2e+10, Einasto bulge: 2.05e+10\n"
     ]
    }
   ],
   "source": [
    "def make_bar_density(rho0, a, b, c, k, n):\n",
    "    def dens_bar(xyz):\n",
    "        # generalized ellipse\n",
    "        r = (abs(xyz[:,0]/a)**k + abs(xyz[:,1]/b)**k + abs(xyz[:,2]/c)**k)**(1./k)\n",
    "        return rho0 * numpy.exp(-r**(1./n))\n",
    "    return dens_bar\n",
    "\n",
    "rho0 = 5e11\n",
    "r0   = 3e-3\n",
    "k    = 2.0\n",
    "n    = 3.0\n",
    "dens_my  = make_bar_density(rho0, r0, r0, r0, k, n)\n",
    "print(dens_my)\n",
    "dens_ein = agama.Density(type='Spheroid', gamma=0, beta=0, outerCutoffRadius=r0, cutoffStrength=1./n, densitynorm=rho0)\n",
    "# compare the density with the spherical Sersic bulge defined earlier\n",
    "pot_bulge = pot_total3[0]  # or agama.Potential(pot_params[0])\n",
    "plt.figure(figsize=(6,4))\n",
    "#plt.loglog(gridR, pot_bulge.density(gridR00), c='r', label='Sersic')  # the spherical Sersic bulge we created earlier\n",
    "plt.loglog(gridR, dens_ein .density(gridR00), c='g', label='Einasto')\n",
    "plt.loglog(gridR, dens_my          (gridR00),  'b+', label='my profile')\n",
    "plt.xlim(gridR[1], max(gridR))\n",
    "plt.ylim(3e3, 3e11)\n",
    "plt.legend(loc='lower left', numpoints=1)\n",
    "plt.xlabel('r [kpc]')\n",
    "plt.ylabel(r'$\\rho [M\\odot/\\mathsf{kpc}^3]$')\n",
    "print(\"Sersic bulge mass: %.3g, Einasto bulge: %.3g\" % (pot_bulge.totalMass(), dens_ein.totalMass()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Having ensured that our model is mathematically equivalent to the Einasto profile for $k=2$, we picked up the parameters of the spherical bulge that approximately matches the Sersic bulge defined earlier. Now time to make it non-spherical and boxy, but we want to keep roughly the same spherically averaged density profile and mass.\n",
    "\n",
    "How do we check the total mass? The `agama.Density` class has a method `totalMass()`, but our model is so far just a lonely Python function that does nothing more than computing the density at the given points. We can wrap it into the `agama.Density` class to enjoy benefits shared by full-fledged built-in models. We need to explicitly specify the symmetry of our density model (in this case, triaxial – first letter is sufficient), because many operations will be more efficient for more symmetric profiles.\n",
    "\n",
    "If we choose the axis ratio $p=y/x$ and $q=z/x$ and want to keep the overall spherical average the same, we need to multiply all three axes by $(pq)^{-1/3}$, so that their geometric mean remains the same."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Boxy model mass: 3.17e+10\n"
     ]
    }
   ],
   "source": [
    "axisYtoX = 0.40\n",
    "axisZtoX = 0.25\n",
    "r0ell    = r0 * (axisYtoX * axisZtoX)**(-1./3)\n",
    "k        = 4.0  # boxy shape\n",
    "dens_my  = make_bar_density(rho0, r0ell, r0ell * axisYtoX, r0ell * axisZtoX, k, n)\n",
    "dens_bar = agama.Density(dens_my, symmetry='t')\n",
    "print(\"Boxy model mass: %.3g\" % dens_bar.totalMass())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Whoops, the mass turned out to be 50% higher than for an ellipsoidal (non-boxy) model, naturally because there is more space inside a boxy ellipse! Let's fix it and also verify that the shape is indeed boxy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Boxy model mass: 2e+10\n"
     ]
    }
   ],
   "source": [
    "rho0 *= (pot_bulge.totalMass() / dens_bar.totalMass())\n",
    "dens_bar = agama.Density(make_bar_density(rho0, r0ell, r0ell * axisYtoX, r0ell * axisZtoX, k, n), symmetry='t')\n",
    "gridx0z = numpy.column_stack((grid2R.reshape(-1), grid2R.reshape(-1)*0, grid2z.reshape(-1)))\n",
    "gridxy0 = numpy.column_stack((grid2R.reshape(-1), grid2z.reshape(-1), 0*grid2z.reshape(-1)))\n",
    "ax = plt.subplots(1, 2, figsize=(13.5,4))[1]\n",
    "ax[0].clabel(ax[0].contour(grid2R, grid2z, numpy.log10(dens_bar.density(gridxy0).reshape(grid2R.shape)),\n",
    "    levels=numpy.linspace(0, 10, 11), cmap='Reds', vmin=3, vmax=10), fmt='$10^%.0f$')\n",
    "ax[1].clabel(ax[1].contour(grid2R, grid2z, numpy.log10(dens_bar.density(gridx0z).reshape(grid2R.shape)),\n",
    "    levels=numpy.linspace(0, 10, 11), cmap='Reds', vmin=3, vmax=10), fmt='$10^%.0f$')\n",
    "ax[0].set_xlim(-0.5, 30)\n",
    "ax[1].set_xlim(-0.5, 30)\n",
    "ax[0].set_xlabel('x [kpc]')\n",
    "ax[0].set_ylabel('y [kpc]')\n",
    "ax[1].set_xlabel('x [kpc]')\n",
    "ax[1].set_ylabel('z [kpc]')\n",
    "ax[0].set_title(r'face-on in-plane density [$M_\\odot/\\mathsf{kpc}^3$]')\n",
    "ax[1].set_title(r'edge-on in-plane density [$M_\\odot/\\mathsf{kpc}^3$]')\n",
    "print(\"Boxy model mass: %.3g\" % dens_bar.totalMass())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Alright, let's now look at the surface density. We would normally use the same method\n",
    "```\n",
    "dens_bar.projectedDensity(...)\n",
    "```\n",
    "as for a built-in model, but if you try this, it will be *very* slow. The reason is, of course, that our function is defined in Python, and the integration along a single line of sight needs to evaluate it hundreds if not thousands of times – every time passing the control flow from the C++ core to the Python user-defined function and back, which isn't fast at all.\n",
    "\n",
    "To get around this problem, and also to make our bar model more powerful in practice, we will construct a `Multipole` potential from it, much in the same way as for any built-in model (in fact, we may pass the original Python function `dens_my` or wrap it into `agama.Density` beforehand – the result is identical).\n",
    "\n",
    "For this rather flattened model, the order of `Multipole` expansion needs to be ramped up to provide sufficient accuracy. By default, the Fourier expansion in $\\phi$ stops at `mmax`=6, and so does the spherical-harmonic expansion in $\\theta$ parametrized by `lmax`. We may certainly increase the latter to something like 10-16, and the cost of potential evaluation is nearly independent of it; however, it scales almost linearly with `mmax` (plus a fixed overhead), so we might need to restrict it to a lower order. Meanwhile, if we were to use a `CylSpline` potential solver instead of `Multipole`, these considerations still apply, but `mmax` above 12 is *considerably* more expensive to construct for a technical reason."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Boxy model mass: 2e+10, Multipole approximation: 2e+10\n"
     ]
    }
   ],
   "source": [
    "pot_bar = agama.Potential(type='Multipole', density=dens_bar, lmax=20, mmax=10)\n",
    "ax = plt.subplots(1, 2, figsize=(13.5,4))[1]\n",
    "for mod, label, cmap in [\n",
    "    [dens_bar, 'original', 'Reds'],\n",
    "    [pot_bar , 'Multipole', 'Blues'] ]:\n",
    "    ax[0].contour(grid2R, grid2z, numpy.log10(mod.density(gridxy0).reshape(grid2R.shape)),\n",
    "        levels=numpy.linspace(0, 10, 11), cmap=cmap, vmin=3, vmax=10)\n",
    "    ax[1].contour(grid2R, grid2z, numpy.log10(mod.density(gridx0z).reshape(grid2R.shape)),\n",
    "        levels=numpy.linspace(0, 10, 11), cmap=cmap, vmin=3, vmax=10)\n",
    "    ax[1].plot([0], [0], color=plt.get_cmap(cmap)(0.5), label=label)\n",
    "ax[0].set_xlabel('x [kpc]')\n",
    "ax[0].set_ylabel('y [kpc]')\n",
    "ax[1].set_xlabel('x [kpc]')\n",
    "ax[1].set_ylabel('z [kpc]')\n",
    "ax[0].set_title(r'face-on in-plane density [$M_\\odot/\\mathsf{kpc}^3$]')\n",
    "ax[1].set_title(r'edge-on in-plane density [$M_\\odot/\\mathsf{kpc}^3$]')\n",
    "ax[1].legend(loc='upper right')\n",
    "print(\"Boxy model mass: %.3g, Multipole approximation: %.3g\" % (dens_bar.totalMass(), pot_bar.totalMass()))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Well, looks tolerable (remember that the potential and forces are better approximated than the density). \n",
    "\n",
    "Let's now look at the circular-velocity curve of this bar and compare it with the spherical Sersic bulge defined earlier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(6,4))\n",
    "plt.plot(gridR, v_circ(pot_bulge, gridR), label='spherical Sersic', c='r')\n",
    "plt.plot(gridR, v_circ(pot_bar  , gridR), label='boxy Einasto', c='b', dashes=[5,4])\n",
    "plt.xlabel('R [kpc]')\n",
    "plt.ylabel(r'$v_\\mathsf{circ}$ [km/s]')\n",
    "plt.legend(loc='lower right')\n",
    "None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Hmm, doesn't quite look identical, despite our efforts to match the sphericalized density profile. It feels like its total mass is higher, but we've checked that it is the same!\n",
    "\n",
    "But this plot is actually misleading, because we computed the \"circular velocity\" from the force along the $x$ axis, which in a triaxial potential is not quite the same as the azimuthally-averaged radial force (the best proxy for the circular velocity).\n",
    "\n",
    "To fix this, we may create another version of \"bar potential\", except that it will be forced to be axisymmetric by keeping `mmax=0`. We can do this either from the original density model, or from the already existing triaxial `Multipole` potential – the result should be very similar (up to the choice of grid nodes, which we left to be decided automatically)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "pot_bar_axi1 = agama.Potential(type='Multipole', density=dens_bar,  lmax=20, mmax=0)   # bar_axi is a kind of oxymoron..\n",
    "pot_bar_axi2 = agama.Potential(type='Multipole', potential=pot_bar, lmax=20, mmax=0)\n",
    "plt.figure(figsize=(6,4))\n",
    "plt.plot(gridR, v_circ(pot_bulge,    gridR), label='spherical Sersic', c='r')\n",
    "plt.plot(gridR, v_circ(pot_bar  ,    gridR), label='boxy Einasto', c='b', dashes=[5,4])\n",
    "plt.plot(gridR, v_circ(pot_bar_axi1, gridR), label='axisym bar 1', c='g')\n",
    "plt.plot(gridR, v_circ(pot_bar_axi2, gridR), label='axisym bar 2', c='m', dashes=[6,5])\n",
    "plt.xlabel('R [kpc]')\n",
    "plt.ylabel(r'$v_\\mathsf{circ}$ [km/s]')\n",
    "plt.legend(loc='lower right')\n",
    "None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now the peak value is still below that of a spherical model (because ours is a flattened axisymmetric one), but at least it approaches it asymptotically at large radii much faster than when we incorrectly evaluate it only along the $x$ axis."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Potentials from an N-body snapshot\n",
    "The general-purpose potential solvers in `Agama` (`Multipole`, `CylSpline`, and another method not mentioned before, `BasisSet`) can be constructed either from analytic density profiles or from N-body snapshots. For instance, one may take particle positions and masses from an N-body simulation, create a smooth approximation of the galaxy potential, and integrate particle orbits in a frozen potential (assuming that the system is reasonably close to stationary; the smooth potential may be rotating, as discussed below), or compute actions for all particles in the snapshot, etc.\n",
    "\n",
    "Naturally, these potential expansions have a finite spatial resolution: for `Multipole`, the radial resolution is determined by the log-spaced grid in spherical radius, and the angular resolution – by the order of spherical-harmonic expansion (roughly speaking, $2\\pi/l_\\mathrm{max}$ for $\\theta$ and $2\\pi/m_\\mathrm{max}$ for $\\phi$); for `CylSpline`, there are separate grids in $R$ and $z$, and the azimuthal resolution is again set by $m_\\mathrm{max}$. While it may be possible to resolve clumps and other substructures at scales above the resolution limit, this is not what these potential expansions are designed for: the goal is to represent the large-scale features of a single well-shaped stellar system and ignore (smooth out) small-scale ones. Moreover, one can explicitly enforce various degrees of symmetry: for instance, setting $m_\\mathrm{max}=0$ will retain only the axisymmetric part of the potential, and one does not need to randomize azimuthal angles of particles to achieve this. This is controlled by the `symmetry='...'` parameter, with possible variants listed in Table 5 of `reference.pdf`, e.g. `spherical`, `axisymmetric`, `triaxial` (the latter assumes that the principal axes are aligned with $x$, $y$, $z$).\n",
    "\n",
    "One thing to keep in mind is that the smooth potential constructed from an N-body system is inherently noisy (although not as much as the \"true\" N-body potential, but still more than a potential expansion created from an analytic density profile). For a given number of particles, there is a practical limit on the resolution (grid size and order of expansion), beyond which the potential will be dominated by discreteness noise and the approximation accuracy will deteriorate rather than improve. The default parameters (25 grid points and order of angular expansion equal to 6) are reasonable in many cases, at least if the snapshot contains $\\gtrsim 10^5$ particles, but may need to be adjusted on a case-to-case basis. By default, the location of the first/last grid nodes are automatically determined from the particle distribution, but this heuristic approach might not give optimal results, and it's recommended to experiment with manually set parameters. For instance, the vertical resolution of `CylSpline` (the smallest grid segment set by `zmin`) should not be much smaller than the scale height of a stellar disk (this also applies to an analytic density profile). And it is very important that the snapshot is centered at origin and (especially for the `CylSpline` case) aligned with principal axes. \n",
    "\n",
    "The density provided by potential expansions constructed from an N-body snapshot may not be sufficiently accurate: being the second derivative of the potential, it is considerably more affected by discreteness noise than the potential itself and its first derivative (acceleration), both of which should be expected to be well approximated.\n",
    "\n",
    "In the following example, we first create an N-body snapshot by sampling our analytic density profile with equal-mass particles with positions randomly drawn from the probability distribution $\\rho(\\boldsymbol{x})$. Note that this snapshot has no velocity information and is therefore not a system in dynamical equilibrium – creating the latter is a much more complex task, which can also be addressed by `Agama`, but that's another story. For the moment, we only use these positions and masses of particles to visually represent their distribution by a scatterplot, and then to construct a `Multipole` potential from a snapshot, which does resemble the analytic one fairly well."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[-9.424342 -9.41712  -0.087892]\n",
      " [-1.84049   0.299798  0.134729]\n",
      " [ 0.274143 -0.051429  0.090701]\n",
      " ...\n",
      " [ 0.353714 -0.1925    0.206431]\n",
      " [ 3.845419  3.884107  0.958336]\n",
      " [ 1.855448 -0.688666  0.7658  ]]\n",
      "[19998.197448 19998.197448 19998.197448 ... 19998.197448 19998.197448 19998.197448]\n",
      "created an array of particle coordinates (shape (1000000, 3)) and masses (shape (1000000,))\n",
      "created a Multipole potential from particles; total mass=2e+10=2e+10\n"
     ]
    }
   ],
   "source": [
    "numpy.set_printoptions(linewidth=100, precision=6, threshold=100, suppress=True)\n",
    "pos, mass = dens_bar.sample(1000000)\n",
    "print(pos)\n",
    "print(mass)\n",
    "print('created an array of particle coordinates (shape %s) and masses (shape %s)' % (pos.shape, mass.shape))\n",
    "pot_bar_nbody = agama.Potential(type='Multipole', particles=(pos, mass), symmetry='triaxial', lmax=20, mmax=10)\n",
    "print('created a Multipole potential from particles; total mass=%.4g=%.4g' %\n",
    "    (numpy.sum(mass), pot_bar_nbody.totalMass()))\n",
    "# plot the snapshot viewed edge-on and the 2d contour plot of projected density\n",
    "ax=plt.subplots(1, 2, figsize=(16,3.75))[1]\n",
    "ax[0].scatter(pos[:,0], pos[:,2], s=1, linewidths=0, edgecolors='none', c='gray', alpha=0.1, rasterized=True)\n",
    "Sigma = pot_bar_nbody.projectedDensity(gridRz, beta=numpy.pi/2).reshape(grid2R.shape)\n",
    "ax[0].contour(grid2R, grid2z, numpy.log10(Sigma), levels=numpy.linspace(5, 10, 11), cmap='Blues', vmin=5, vmax=10)\n",
    "ax[0].set_xlim(-6, 6)\n",
    "ax[0].set_ylim(-3, 3)\n",
    "ax[0].set_xlabel('X [kpc]')\n",
    "ax[0].set_ylabel('Z [kpc]')\n",
    "ax[0].set_title(r'surface density  [$M_\\odot/\\mathsf{kpc}^2$]')\n",
    "# plot the density in the y=0 plane (original analytic profile, its Multipole approximation,\n",
    "# and the Multipole constructed from an N-body snapshot)\n",
    "for mod, label, cmap, linestyle in [\n",
    "    [dens_bar, 'analytic original', 'Reds', '-'],\n",
    "    [pot_bar , 'analytic Multipole', 'Blues', '-'],\n",
    "    [pot_bar_nbody, 'N-body Multipole', 'Greens', '--']]:\n",
    "    ax[1].contour(grid2R, grid2z, numpy.log10(mod.density(gridx0z).reshape(grid2R.shape)),\n",
    "        levels=numpy.linspace(0, 10, 11), cmap=cmap, vmin=3, vmax=10, linestyles=linestyle)\n",
    "    ax[1].plot([0], [0], color=plt.get_cmap(cmap)(0.5), label=label)\n",
    "ax[1].set_xlim(0, 24)\n",
    "ax[1].set_ylim(-6, 6)\n",
    "ax[1].set_xlabel('x [kpc]')\n",
    "ax[1].set_ylabel('z [kpc]')\n",
    "ax[1].set_title(r'in-plane density [$M_\\odot/\\mathsf{kpc}^3$]')\n",
    "ax[1].legend(loc='upper right')\n",
    "None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In fact, the density of the two `Multipole` expansions (created from the original analytic profile or from its N-body representation) are so similar as to be nearly indistinguishable; clearly, in this case the discreteness noise is still sub-dominant to the errors caused by a finite order of expansion. We can quantify it by computing the distribution of relative errors in density: the average error is slightly larger for the N-body case, but the tail of higher errors is nearly the same (and these appear only in the very-low-density outer regions)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "dens_true = dens_bar.density(pos[::100])\n",
    "dens_multipole_analytic = pot_bar.density(pos[::100])\n",
    "dens_multipole_nbody = pot_bar_nbody.density(pos[::100])\n",
    "err_analytic = abs(dens_multipole_analytic / dens_true - 1)\n",
    "err_nbody    = abs(dens_multipole_nbody    / dens_true - 1)\n",
    "err_bins = numpy.logspace(-4,0,41)\n",
    "plt.figure(figsize=(6,4))\n",
    "plt.hist(err_analytic, bins=err_bins, histtype='step', label='analytic Multipole', color='b')\n",
    "plt.hist(err_nbody,    bins=err_bins, histtype='step', label='N-body Multipole', color='r')\n",
    "plt.legend(loc='upper left')\n",
    "plt.xlabel('relative error in density')\n",
    "plt.gca().set_yticklabels([])\n",
    "plt.xscale('log')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Density/Potential modifiers\n",
    "All potentials considered so far were static, which is the most common usage scenario.\n",
    "Even in this case, orbits can be computed in a rotating reference frame, as we will discuss shortly.\n",
    "\n",
    "But sometimes one needs to consider a system of two galaxies moving w.r.t. each other, or a bar rotating with a gradually decreasing pattern speed and/or increasing in amplitude, etc. \n",
    "This is achieved by adding \"modifiers\" either to existing density or potential objects or at the time of construction.  \n",
    "See `example_time_dependent_potential.py` for an illustration of various approaches for creating time-dependent potentials.\n",
    "\n",
    "For example, let us create a full Milky Way potential (the boxy bar and the two other components introduced earlier) which rotates with a constant pattern speed $\\Omega_p$ (later we will make it variable).\n",
    "The `Rotating` modifier rotates the potential figure about the $z$ axis by an angle that is has an arbitrary dependence on time (represented by a cubic spline function), and of course, a constant pattern speed corresponds to a linearly increasing (or decreasing) angle, so needs only two points to specify it. The timestamps and corresponding angles are provided as a 2d array with two columns (either directly to the constructor, or first writing the array into a file and providing its name). We create a new instance of `Potential` from a static potential that is wrapped into the modifier (the original static potential remains unchanged) and specifying the modifier parameters in the `rotation=...` argument. Alternatively, if we were to create a new potential with the given parameters (`type=...`, `mass=...`, etc.), we could provide `rotation=...` directly in the list of arguments to the constructor, instead of a two-stage procedure used below.\n",
    "\n",
    "For the case of the Milky Way, we choose $\\Omega_p=-40$ km/s/kpc (negative since it rotates clockwise in the $x-y$ plane for the standard choice of right-handed Galactocentric coordinate system, in which the Sun is at $x=-8$ kpc)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CompositePotential{ Multipole, CompositePotential{ DiskAnsatz, Multipole }, Multipole } (symmetry: Triaxial)\n",
      "Rotating CompositePotential{ Multipole, CompositePotential{ DiskAnsatz, Multipole }, Multipole } (symmetry: Bisymmetric)\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "pot_mw_static = agama.Potential(pot_bar, pot_total3[1], pot_total3[2])\n",
    "# create a file with the rotation angle as a function of time\n",
    "Omega_p = -40.0  # in our units, km/s/kpc\n",
    "pot_mw_rotating = agama.Potential(potential=pot_mw_static, rotation=[[0, 0], [1, Omega_p]])\n",
    "print(pot_mw_static)\n",
    "print(pot_mw_rotating)\n",
    "# verify that the underlying potential wrapped into the Rotating modifier is the same as input one\n",
    "print(pot_mw_rotating[0] == pot_mw_static)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Orbit integration\n",
    "One of the common tasks associated with `Potential` objects is numerical orbit integration, which is performed by a standalone routine `agama.orbit()` (not a method of `Potential` class). It has several required parameters: initial conditions for one or more orbits (`ic`), the instance of a `potential`, total integration `time` (which may be the same for all orbits or specified individually), and a number of optional ones – in our context, the size of output trajectory (again may be specified separately for each orbit). In addition to the trajectory itself, it can return other quantities associated with an orbit, but these are for another story.\n",
    "\n",
    "The trajectory can be returned by the orbit integration routine in two alternative formats: as an array of timestamps and 6d phase-space points, or as an interpolator. \n",
    " - In the \\[historically\\] first case, each orbit produces two arrays: timestamps (1d array of length `trajsize`) and trajectory (2d array with `trajsize` rows and 6 columns – position and velocity in cartesian coordinates). Because the trajectory sizes can differ between orbits, in general one cannot represent the trajectories as a 3d `numpy` array of shape `(num_orbits, trajsize, 6)`; instead, the result is a 2d array of shape `(num_orbits, 2)`, in which each element is an array itself: 0th column contains timestamps and 1st column – trajectories. This is illustrated below by printing out the result array and its elements. \n",
    " - In the second case, each orbit is represented by an instance of a special `Orbit` class, which is a thin layer on top of three quintic `Spline` interpolators for each cartesian component. This object provides a function interface, computing the interpolated orbit at any given time (or an array of times in a single call), and a sequence interface, allowing one to iterate over the timestamps at which an orbit was recorded. Combining the two interfaces, one may retrieve the positions/velocities at the corresponding timestamps by a rather funny-looking application of the Orbit object to itself.\n",
    "\n",
    "In fact, the `orbit()` routine can also perform integration in the rotating frame, without the need to create a rotating potential, but only for the case of a constant pattern speed. So we will do it twice, first with `pot_mw_static` and providing $\\Omega_p$ to the `orbit()` routine, second with `pot_mw_rotating`.\n",
    "\n",
    "The bar in our potential is aligned with the X axis, and we consider two orbits with initial conditions at the Solar location, which is trailing behind the bar (offset by $25^\\circ$), and with different values of tangential velocity."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Initial conditions (first aligned with X axis, then rotated):\n",
      "[[ -8.         0.         0.         0.       215.2371     0.      ]\n",
      " [ -8.         0.         0.         0.       191.321867   0.      ]]\n",
      "[[ -7.250462  -3.380946   0.       -90.963129 195.07106    0.      ]\n",
      " [ -7.250462  -3.380946   0.       -80.856115 173.396498   0.      ]]\n",
      "\n",
      "2 orbits complete (285.7 orbits/s)\n",
      "Result of orbit integration represented by Orbit interpolators (array of shape (2,)):\n",
      "[Orbit with 467 nodes on t=[0:6] Orbit with 679 nodes on t=[0:6]]\n",
      "0th orbit times (length 467) and trajectory (shape (467, 6)):\n",
      "[0.       0.001963 0.01374  ... 5.982757 5.995217 6.      ]\n",
      "[[  -7.250462   -3.380946    0.        -90.963129  195.07106     0.      ]\n",
      " [  -7.159387   -3.564456    0.        -94.117662  194.408415    0.      ]\n",
      " [  -6.47756    -4.514806    0.       -115.964158  190.952971    0.      ]\n",
      " ...\n",
      " [   7.361554    0.708182    0.         -0.530364 -220.068539    0.      ]\n",
      " [   6.893345    1.44362     0.         14.970227 -234.197403    0.      ]\n",
      " [   6.688999    1.607327    0.         22.200095 -240.64472     0.      ]]\n",
      "\n",
      "2 orbits complete (333.3 orbits/s)\n",
      "Result of orbit integration represented by timestamps and trajectories (array of shape (2, 2)):\n",
      "[[array([0.  , 0.01, 0.02, ..., 5.98, 5.99, 6.  ], dtype=float32)\n",
      "  array([[  -7.250463,   -3.380946,    0.      ,  -90.96313 ,  195.07106 ,    0.      ],\n",
      "         [  -7.836354,   -1.29485 ,    0.      ,  -25.139576,  219.16927 ,    0.      ],\n",
      "         [  -7.738189,    0.938794,    0.      ,   45.15441 ,  224.18384 ,    0.      ],\n",
      "         ...,\n",
      "         [   6.939559,   -2.710342,    0.      ,  -95.67527 , -195.76529 ,    0.      ],\n",
      "         [   5.63384 ,   -4.496447,    0.      , -164.03854 , -157.98395 ,    0.      ],\n",
      "         [   3.698847,   -5.800427,    0.      , -220.28302 ,  -99.38835 ,    0.      ]],\n",
      "        dtype=float32)                                                                   ]\n",
      " [array([0.  , 0.01, 0.02, ..., 5.98, 5.99, 6.  ], dtype=float32)\n",
      "  array([[  -7.250463,   -3.380946,    0.      ,  -80.85612 ,  173.3965  ,    0.      ],\n",
      "         [  -7.736289,   -1.508234,    0.      ,  -15.238904,  198.51363 ,    0.      ],\n",
      "         [  -7.5363  ,    0.532035,    0.      ,   56.012024,  206.3973  ,    0.      ],\n",
      "         ...,\n",
      "         [  -7.26868 ,   -5.060413,    0.      , -138.61057 ,  142.7678  ,    0.      ],\n",
      "         [  -8.400467,   -3.475266,    0.      ,  -86.9476  ,  172.16042 ,    0.      ],\n",
      "         [  -8.994542,   -1.658567,    0.      ,  -31.338728,  189.12799 ,    0.      ]],\n",
      "        dtype=float32)                                                                   ]]\n",
      "0th orbit: timestamps (shape (601,)) and trajectory (shape (601, 6)):\n",
      "[0.   0.01 0.02 ... 5.98 5.99 6.  ]\n",
      "[[  -7.250463   -3.380946    0.        -90.96313   195.07106     0.      ]\n",
      " [  -7.836354   -1.29485     0.        -25.139576  219.16927     0.      ]\n",
      " [  -7.738189    0.938794    0.         45.15441   224.18384     0.      ]\n",
      " ...\n",
      " [   6.939559   -2.710342    0.        -95.67527  -195.76529     0.      ]\n",
      " [   5.63384    -4.496447    0.       -164.03854  -157.98395     0.      ]\n",
      " [   3.698847   -5.800427    0.       -220.28302   -99.38835     0.      ]]\n"
     ]
    }
   ],
   "source": [
    "# auxiliary routine for rotating an array of position/velocity points by a given angle\n",
    "def rotate(posvel, angle):\n",
    "    x, y, z, vx, vy, vz = posvel.T\n",
    "    sina, cosa = numpy.sin(angle), numpy.cos(angle)\n",
    "    return numpy.array([x*cosa-y*sina, x*sina+y*cosa, z, vx*cosa-vy*sina, vx*sina+vy*cosa, vz]).T\n",
    "\n",
    "bar_angle = 25.0 * numpy.pi/180\n",
    "R0 = 8.0\n",
    "vcirc0 = v_circ(pot_mw_static, R0)\n",
    "ictmp = numpy.array([\n",
    "    [-R0, 0, 0, 0.0, 0.9*vcirc0, 0],\n",
    "    [-R0, 0, 0, 0.0, 0.8*vcirc0, 0] ])  # temporarily place IC along the X axis\n",
    "ic = rotate(ictmp, bar_angle)           # and then rotate to place it below the X axis\n",
    "print('Initial conditions (first aligned with X axis, then rotated):\\n%s\\n%s\\n' % (ictmp, ic))\n",
    "\n",
    "inttime  = 6.0  # expressed in our time units, i.e. 0.98 Gyr - in what follows, we approximately refer to it as Gyr\n",
    "# integrate the orbits in the rotating frame, using a static potential and providing the pattern speed\n",
    "result1 = agama.orbit(potential=pot_mw_static, ic=ic, time=inttime, dtype=object, Omega=Omega_p)\n",
    "print('Result of orbit integration represented by Orbit interpolators (array of shape %s):\\n%s' %\n",
    "    (result1.shape, result1))\n",
    "# convert the orbit interpolator into an array of times and 6d trajectory points.\n",
    "# in this expression you're not accidentally seeing double:\n",
    "# the 'orbit' object inside the round brackets produces the array of timestamps,\n",
    "# while the same 'orbit' object acting as a function on the array of timestamps\n",
    "# produces the corresponding 6d phase-space points;\n",
    "# of course, one may evaluate it at any other time between orbit[0]=0 and orbit[trajsize-1]=inttime\n",
    "trajs1 = [orbit(orbit) for orbit in result1]\n",
    "numpy.set_printoptions(threshold=100, linewidth=100, edgeitems=3)   # reduce the size of the printout\n",
    "print('0th orbit times (length %i) and trajectory (shape %s):\\n%s\\n%s\\n' %\n",
    "    (len(result1[0]), trajs1[0].shape, numpy.array(result1[0]), trajs1[0]))\n",
    "\n",
    "# integrate the orbits in the inertial frame, using a rotating potential\n",
    "trajsize = 601  # this time we output the orbits as arrays at regularly spaced times\n",
    "result2 = agama.orbit(potential=pot_mw_rotating, ic=ic, time=inttime, trajsize=trajsize)\n",
    "print('Result of orbit integration represented by timestamps and trajectories (array of shape %s):\\n%s' %\n",
    "    (result2.shape, result2))\n",
    "print('0th orbit: timestamps (shape %s) and trajectory (shape %s):\\n%s\\n%s' %\n",
    "    (result2[0,0].shape, result2[0,1].shape, result2[0,0], result2[0,1]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "When integrating an orbit in the rotating frame, the initial conditions and the output trajectory are specified in the sequence of inertial reference frames that instantaneously coincide with the rotating frame at all times. This might look somewhat counter-intuitive, but actually makes common sense: in the real Milky Way, we measure the positions and velocities of stars in the present-day inertial reference frame, and these values are independent of the assumed pattern speed. That's why the positions in the returned are specified in the bar-aligned frames, but for each moment of time, the velocities refer to the inertial frame that coincides with the rotating frame at this moment. In other words, the time derivative of position is not velocity, but actually $\\dot{\\boldsymbol x} = {\\boldsymbol v} - \\boldsymbol{\\Omega}\\times \\boldsymbol{x}$.\n",
    "\n",
    "On the other hand, when integrating the orbits in an inertial frame and using the rotating potential, the resulting trajectories will be specified in the same inertial frame at all times, instead of being given in a sequence of _different_ inertial frames that coincide with the rotating frame at each moment. So we will need to rotate the resulting trajectories to align the positions with the rotating frame in which the bar always has the same orientation, and also rotate the velocities correspondingly.\n",
    "\n",
    "The two orbits have quite different appearance: one is a \"normal\" circulating orbit, the other is a \"horseshoe\" resonantly-trapped orbit."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "ax = plt.subplots(1, 3, figsize=(10,4), sharex=True, sharey=True)[1]\n",
    "for traj, color in zip(trajs1, ['b','r']):\n",
    "    ax[0].plot(traj[:,0], traj[:,1], color=color)\n",
    "ax[0].set_title('integrated in rotating frame')\n",
    "for (t, traj), color in zip(result2, ['b','r']):\n",
    "    ax[1].plot(traj[:,0], traj[:,1], color=color)\n",
    "    rot_traj = rotate(traj, -t*Omega_p)\n",
    "    ax[2].plot(rot_traj[:,0], rot_traj[:,1], color=color)\n",
    "ax[1].set_title('integrated in inertial frame')\n",
    "ax[2].set_title('same but rotated afterwards')\n",
    "for a in ax:\n",
    "    a.set_aspect('equal')\n",
    "    a.set_xlabel('x [kpc]')\n",
    "    a.set_ylabel('y [kpc]')\n",
    "    a.plot(ic[0,0], ic[0,1], 'ko')  # initial point, common for both orbits"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now comes the fun part: when using the time-dependent rotating potential, we may make the pattern speed vary with time. The rotation angle is specified as a cubic spline, so the pattern speed will be a piecewise-quadratic function, and in this example we make it smoothly change it from the initial value $\\Omega_p$ for times $t\\le t_p$ to a lower final value $\\Omega_f$ at times $t\\ge t_f$. We need to add an intermediate timestamp halfway between $t_p$ and $t_m$ in order to \"bend\" the spline (with only two points, it would be a linear function). To illustrate the time evolution of the rotation angle and its derivative (pattern speed), we plot both against time; note that the pattern speed is extrapolated as a constant beyond both endpoints."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cubic spline with 3 nodes on [0.5:5.5]\n"
     ]
    }
   ],
   "source": [
    "Omega_f = Omega_p * 0.5\n",
    "t_p = 0.5  # time at which the potential starts to slow down, again in our time units\n",
    "t_f = 5.5  # time at which it achieves the final pattern speed\n",
    "rot_angle = numpy.array([\n",
    "    [ t_p,             t_p*Omega_p],\n",
    "    [(t_p + t_f) / 2., t_p*Omega_p + (t_f-t_p) * (5*Omega_p + Omega_f) / 12.],\n",
    "    [ t_f,             t_p*Omega_p + (t_f-t_p) * (  Omega_p + Omega_f) / 2. ],\n",
    "])\n",
    "rot_angle_spl = agama.Spline(rot_angle[:,0], rot_angle[:,1])\n",
    "print(rot_angle_spl)\n",
    "ax = plt.subplots(1, 2, figsize=(10,4))[1]\n",
    "tt = numpy.linspace(-2, 8, 201)\n",
    "ax[0].plot(tt, rot_angle_spl(tt), 'b')\n",
    "ax[1].plot(tt, rot_angle_spl(tt, der=1), 'b')\n",
    "ax[0].plot(rot_angle[:,0], rot_angle[:,1], 'ko')\n",
    "ax[1].plot(rot_angle[:,0], rot_angle_spl(rot_angle[:,0], der=1), 'ko')\n",
    "ax[0].yaxis.set_major_locator(matplotlib.ticker.MultipleLocator(10*numpy.pi))\n",
    "ax[0].yaxis.set_major_formatter(matplotlib.ticker.FuncFormatter(\n",
    "    lambda val,pos: r'$%.0f\\pi$' % (val/numpy.pi) if val !=0 else '0'))\n",
    "ax[0].set_xlabel('Time [Gyr]')\n",
    "ax[0].set_ylabel('rotation angle [radians]')\n",
    "ax[1].set_xlabel('Time [Gyr]')\n",
    "ax[1].set_ylabel(r'pattern speed $\\Omega$ [km/s/kpc]')\n",
    "ax[1].set_ylim(-42,-18)\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We now create another instance of rotating potential with the provided table of rotation angles, integrate two orbits with the same initial conditions, and transform the trajectories into the rotating frame (using the spline function constructed above to evaluate the rotation angle at any moment of time). \n",
    "\n",
    "Note how the orbit trapped at the horseshoe resonance is pushed out by the slowing-down bar, and the other orbit also changed its character! We can see it more clearly in the right panel, which plots the time evolution of the $y$ coordinate for both steadily rotating and slowing-down potentials. These coincide at early times ($t<t_p$) when the pattern speed stayed at the original value, and then start to diverge."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2 orbits complete (285.7 orbits/s)\n"
     ]
    }
   ],
   "source": [
    "pot_mw_slowing = agama.Potential(potential=pot_mw_static, rotation=rot_angle)\n",
    "\n",
    "result3 = agama.orbit(potential=pot_mw_slowing, ic=ic, time=inttime, trajsize=trajsize)\n",
    "ax = plt.subplots(1, 3, figsize=(10,4), sharey=True)[1]\n",
    "for t, traj, color in zip(result1, trajs1, ['b','r']):\n",
    "    ax[0].plot(traj[:,0], traj[:,1], color=color)\n",
    "    #ax[1].plot(numpy.nan, numpy.nan)  # dummy curve to \"steal\" the color from the standard cycle\n",
    "    ax[2].plot(t, traj[:,1], color=color)\n",
    "ax[0].set_aspect('equal')\n",
    "ax[0].set_title('steadily rotating potential')\n",
    "for (t, traj), color in zip(result3, ['c','m']):\n",
    "    rot_traj = rotate(traj, -rot_angle_spl(t))\n",
    "    ax[1].plot(rot_traj[:,0], rot_traj[:,1], color=color)\n",
    "    ax[2].plot(t, rot_traj[:,1], color=color)\n",
    "ax[1].set_aspect('equal')\n",
    "ax[1].set_title('slowing down potential')\n",
    "ax[2].set_title('time evolution of y coordinate')\n",
    "ax[0].plot(ic[0,0], ic[0,1], 'ko')  # initial point, common for both orbits\n",
    "ax[1].plot(ic[0,0], ic[0,1], 'ko')\n",
    "ax[0].set_xlabel('x [kpc]')\n",
    "ax[0].set_ylabel('y [kpc]')\n",
    "ax[1].set_xlabel('x [kpc]')\n",
    "ax[2].set_xlabel('time [Gyr]')\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We may also wish to change the bar properties as it slows down, for instance, increase its length and mass. This can be achieved by the `Scaled` modifier applied to the bar potential.\n",
    "\n",
    "The `Scaled` modifier similarly reads its parameters from a 2d array or from a text file, where the mass normalization and the length multiplier are given as functions of time, and constructs spline interpolators for both of them. So the file should contain at least three columns (timestamp, mass scale and length scale). Since the splines will be linearly extrapolated beyond endpoints of the time interval, using the derivatives taken at the corresponding point, we need to take extra effort to ensure that the extrapolation actually gives a constant (i.e., the derivative is zero). There are two alternative ways to achieve this:\n",
    "\n",
    " - add another timestamp beyond the endpoint of the original time interval, and specify the same value at two adjacent points: in this case, a regularized spline will have a constant value at the corresponding grid segment, and therefore will be extrapolated as a constant;\n",
    " - manually specify not only the values, but also the derivatives of our functions at each timestamp: in this case, a so-called Hermite spline will be created with prescribed derivatives at each point.\n",
    "\n",
    "We illustrate both approaches, creating corresponding `agama.Spline` objects, which have the same properties as the cubic splines internally constructed by the `Scaled` modifier, and comparing the value of the time-dependent potential at different times and correspondingly scaled positions."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "M_f = 2.0  # multiplier for the bar mass at the final time t_f\n",
    "R_f = 1.2  # multiplier for the radius at t_f\n",
    "\n",
    "# first approach: create extra points with duplicate values\n",
    "bar_scale1 = numpy.array([\n",
    "    [0.0, 1.0, 1.0],\n",
    "    [t_p, 1.0, 1.0],\n",
    "    [t_f, M_f, R_f],\n",
    "    [999, M_f, R_f]\n",
    "])\n",
    "mass_scale_spl1 = agama.Spline(bar_scale1[:,0], bar_scale1[:,1], reg=True)  # regularized spline,\n",
    "size_scale_spl1 = agama.Spline(bar_scale1[:,0], bar_scale1[:,2], reg=True)  # same as in the modifier\n",
    "# second approach: manually specify zero derivatives at endpoints\n",
    "bar_scale2 = numpy.array([\n",
    "    [t_p, 1.0, 1.0, 0, 0],\n",
    "    [t_f, M_f, R_f, 0, 0],\n",
    "])\n",
    "mass_scale_spl2 = agama.Spline(bar_scale2[:,0], bar_scale2[:,1], left=0, right=0)\n",
    "size_scale_spl2 = agama.Spline(bar_scale2[:,0], bar_scale2[:,2], left=0, right=0)\n",
    "\n",
    "pot_bar_slowing_growing1 = agama.Potential(potential=pot_bar, scale=bar_scale1, rotation=rot_angle)\n",
    "pot_bar_slowing_growing2 = agama.Potential(potential=pot_bar, scale=bar_scale2, rotation=rot_angle)\n",
    "\n",
    "# verify that the time-dependent potential is proportional to mass scale divided by size scale at all times\n",
    "tt = numpy.linspace(-2.0, 8.0, 101)\n",
    "point = numpy.array([0, 0, 1.0])  # evaluate along the z axis to avoid complications associated with rotation\n",
    "pot_bar_t0 = pot_bar.potential(point)\n",
    "# scale the point coordinate by the size function at each moment of time\n",
    "ax = plt.subplots(1, 2, figsize=(8,4), sharex=True, sharey=True)[1]\n",
    "ax[0].plot(tt, [pot_bar_slowing_growing1.potential(point * size_scale_spl1(t), t=t) / pot_bar_t0 for t in tt], color='b')\n",
    "ax[1].plot(tt, [pot_bar_slowing_growing2.potential(point * size_scale_spl2(t), t=t) / pot_bar_t0 for t in tt], color='b')\n",
    "ax[0].plot(tt, mass_scale_spl1(tt) / size_scale_spl1(tt), 'r--')\n",
    "ax[1].plot(tt, mass_scale_spl2(tt) / size_scale_spl2(tt), 'r--')  # the curves should be indistinguishable\n",
    "ax[0].set_xlabel('time [Gyr]')\n",
    "ax[1].set_xlabel('time [Gyr]')\n",
    "ax[0].set_ylabel('potential normalization')\n",
    "None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Since our scaled and rotating bar is the only non-axisymmetric component, we add the disk and the halo components without any modifications, and compute our two orbits in this total potential. They do look different from the previous cases.."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CompositePotential{ Rotating Scaled Multipole, CompositePotential{ DiskAnsatz, Multipole }, Multipole } (symmetry: Bisymmetric)\n",
      "2 orbits complete (200 orbits/s)\n"
     ]
    }
   ],
   "source": [
    "pot_mw_slowing_growing = agama.Potential(pot_bar_slowing_growing1, pot_total3[1], pot_total3[2])\n",
    "print(pot_mw_slowing_growing)\n",
    "\n",
    "result4 = agama.orbit(potential=pot_mw_slowing_growing, ic=ic, time=inttime, trajsize=trajsize)\n",
    "\n",
    "ax = plt.subplots(1, 3, figsize=(10,4), sharey=True)[1]\n",
    "for traj in trajs1:\n",
    "    ax[0].plot(traj[:,0], traj[:,1])\n",
    "    ax[1].plot(numpy.nan, numpy.nan)  # dummy curve to \"steal\" the color from the standard cycle\n",
    "    ax[2].plot(numpy.nan, [[numpy.nan, numpy.nan]])\n",
    "ax[0].set_aspect('equal')\n",
    "ax[0].set_title('steadily rotating potential')\n",
    "for t,traj in result3:\n",
    "    rot_traj = rotate(traj, -rot_angle_spl(t))\n",
    "    ax[1].plot(rot_traj[:,0], rot_traj[:,1])\n",
    "ax[1].set_aspect('equal')\n",
    "ax[1].set_title('slowing down potential')\n",
    "for t,traj in result4:\n",
    "    rot_traj = rotate(traj, -rot_angle_spl(t))\n",
    "    ax[2].plot(rot_traj[:,0], rot_traj[:,1])\n",
    "ax[2].set_aspect('equal')\n",
    "ax[2].set_title('slowing and growing potential')\n",
    "ax[0].set_xlabel('x [kpc]')\n",
    "ax[0].set_ylabel('y [kpc]')\n",
    "ax[1].set_xlabel('x [kpc]')\n",
    "ax[2].set_xlabel('x [kpc]')\n",
    "plt.tight_layout()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we illustrate the changes imparted on the orbits in an evolving potential by plotting the time evolution of the Jacobi energy $J = \\Phi(\\boldsymbol x) + |\\boldsymbol v|^2/2 - \\Omega\\,L_z$. In the case of a steadily rotating potential, this is an integral of motion (generalization of the energy integral), but in a time-dependent potential, all bets are off, and we see that the Jacobi integral decreases in magnitude, as if the orbits were pushed out (which is indeed the case for the horseshoe orbit, but the other orbit becomes more compact in space and instead increases its kinetic energy)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "def JacobiEnergy(pot, t, traj, Omega):\n",
    "    Lz = traj[:,0] * traj[:,4] - traj[:,1] * traj[:,3]\n",
    "    # note that the potential is evaluated at the timestamps corresponding to each point on the trajectory\n",
    "    return pot.potential(traj[:,0:3], t=t) + 0.5 * numpy.sum(traj[:,3:6]**2, axis=1) - Omega * Lz\n",
    "\n",
    "plt.figure(figsize=(6,4))\n",
    "for t,traj in result2:\n",
    "    plt.plot(t, JacobiEnergy(pot_mw_rotating, t, traj, Omega_p), label='constant rotation')\n",
    "for t,traj in result3:\n",
    "    plt.plot(t, JacobiEnergy(pot_mw_slowing, t, traj, rot_angle_spl(t, der=1)), label='slowing down')\n",
    "for t,traj in result4:\n",
    "    plt.plot(t, JacobiEnergy(pot_mw_slowing_growing, t, traj, rot_angle_spl(t, der=1)), label='slowing and growing')\n",
    "\n",
    "plt.xlabel('time [Gyr]')\n",
    "plt.ylabel('Jacobi energy [(km/s)$^2$]')\n",
    "plt.legend(loc='upper left')\n",
    "None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
